{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f6ae87d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#These are scripts which produce plots which originated in python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a44219a7",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pandas'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Input \u001b[0;32mIn [2]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m### plotting Fig 1C\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mos\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01msys\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pandas'"
     ]
    }
   ],
   "source": [
    "### plotting Fig 1C\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "from PIL import Image\n",
    "import math\n",
    "import numpy as np\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "import statsmodels.stats.multitest as multi\n",
    "\n",
    "table = 'processed_data/benchmark_GM12878_results.txt'\n",
    "df = pd.read_table(table)\n",
    "\n",
    "#### filtering out negative controls with significance\n",
    "df = df[~df['S'].str.contains('random') | (df['S'].str.contains('random') & (df['padj_S_run1'] > 0.001))]\n",
    "df = df[~df['S'].str.contains('random') | (df['S'].str.contains('random') & (df['padj_S_run2'] > 0.001))]\n",
    "\n",
    "## calc Pearson's r\n",
    "df = df.drop_duplicates(subset=['log2FC_ES'])\n",
    "df1 = df.dropna(subset=['log2FC_ES', 'log2FC_SE'])\n",
    "cor = df1.iloc[:,[8,14]]\n",
    "cor_p = cor.corr(method='pearson')\n",
    "cor_p = cor_p.iloc[0,1]\n",
    "\n",
    "r1 = 'r = '+ '{:.2f}'.format(cor_p)\n",
    "\n",
    "## plot\n",
    "fig = plt.figure(figsize=(4,4), dpi=300)\n",
    "ax = fig.add_subplot(111, xticks=[-2,0,2,4,6],yticks=[-2,0,2,4,6])\n",
    "plt.xlim((-2.9,7.5))\n",
    "plt.ylim((-2.9,7.5))\n",
    "plt.axhline(y=0, color='lightgrey', linestyle='dotted', zorder=0)\n",
    "plt.axvline(x=0, color='lightgrey', linestyle='dotted', zorder=0)\n",
    "\n",
    "## scatter plot\n",
    "plt.scatter(y=df['log2FC_ES'], x=df['log2FC_SE'], s=1, alpha=0.1,color='0.3', zorder=1)\n",
    "\n",
    "## 45 degree line\n",
    "x2 = np.arange(-10,10,0.1)\n",
    "y2 = x2\n",
    "plt.plot(x2,y2, color='coral', alpha=0.8, zorder=4, linestyle='dotted')\n",
    "plt.text(-2.3,6.5,r1, color='0.2')\n",
    "\n",
    "plt.ylabel('Expression log2 (Fold Change)', color='0.2')\n",
    "plt.xlabel('Expression log2 (Fold Change)', color='0.2')\n",
    "\n",
    "k= '0.4'\n",
    "ax.spines['right'].set_color(k)\n",
    "ax.spines['left'].set_color(k)\n",
    "ax.spines['top'].set_color(k)\n",
    "ax.spines['bottom'].set_color(k)\n",
    "ax.tick_params(axis='x', colors=k)\n",
    "ax.tick_params(axis='y', colors=k)\n",
    "\n",
    "plt.savefig('Fig1C.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73de21ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "### plotting Fig 1D\n",
    "### plotting duo and log additive model\n",
    "\n",
    "df = pd.read_table(table)\n",
    "\n",
    "#### filtering out negative controls with significance\n",
    "df = df[~df['S'].str.contains('random') | (df['S'].str.contains('random') & (df['padj_S_run1'] > 0.001))]\n",
    "df = df[~df['S'].str.contains('random') | (df['S'].str.contains('random') & (df['padj_S_run2'] > 0.001))]\n",
    "\n",
    "df = df.dropna(subset=['log2FC_ES', 'log2FC_SE'])\n",
    "\n",
    "# making single x single term\n",
    "df['ES_ExS'] = df['log2FC_S_run2']*df['log2FC_E_run2']\n",
    "df['SE_ExS'] = df['log2FC_S_run1']*df['log2FC_E_run1']\n",
    "\n",
    "# linear regression\n",
    "run2 = df.loc[:,['log2FC_S_run1', 'log2FC_E_run1','SE_ExS']].as_matrix()\n",
    "SE = df['log2FC_SE'].as_matrix()\n",
    "run3 = df.loc[:,['log2FC_S_run2', 'log2FC_E_run2', 'ES_ExS']].as_matrix()\n",
    "ES = df['log2FC_ES'].as_matrix()\n",
    "\n",
    "linSE = lm.LinearRegression()\n",
    "linES = lm.LinearRegression()\n",
    "linSE.fit(run2,SE)\n",
    "linES.fit(run3,ES)\n",
    "\n",
    "SEscore = linSE.score(run2,SE)\n",
    "ESscore = linES.score(run3,ES)\n",
    "print linSE.coef_\n",
    "print linES.coef_\n",
    "\n",
    "## add a column of additive model to plot\n",
    "df['AP_SE_fit'] = linSE.intercept_ + (df['log2FC_S_run1'] * linSE.coef_[0]) + (df['log2FC_E_run1']*linSE.coef_[1]) + (df['SE_ExS']*linSE.coef_[2])\n",
    "df['AP_ES_fit'] = linES.intercept_ + (df['log2FC_S_run2'] * linES.coef_[0]) + (df['log2FC_E_run2']*linES.coef_[1]) + (df['ES_ExS']*linSE.coef_[2])\n",
    "\n",
    "\n",
    "SEscore = 'r^2 = '+ '{:.3f}'.format(linSE.score(run2,SE))\n",
    "ESscore = 'r^2 = '+ '{:.3f}'.format(linES.score(run3,ES))\n",
    "SEfunc = 'SE = '+'{:.3f}'.format(linSE.intercept_)+' + '+'{:.3f}'.format(linSE.coef_[0])+'S + '+'{:.3f}'.format(linSE.coef_[1])+'E + '+'{:.3f}'.format(linSE.coef_[2])+'ExS'\n",
    "ESfunc = 'ES = '+'{:.3f}'.format(linES.intercept_)+' + '+'{:.3f}'.format(linES.coef_[0])+'S + '+'{:.3f}'.format(linES.coef_[1])+'E + '+'{:.3f}'.format(linES.coef_[2])+'ExS'\n",
    "\n",
    "for key in ['ES', 'SE']:\n",
    "    x = 'AP_%s_fit' % key\n",
    "    y = 'log2FC_%s' % key\n",
    "\n",
    "    ## plot\n",
    "    fig = plt.figure(figsize=(4,4), dpi=300)\n",
    "    ax = fig.add_subplot(111, xticks=[-2,0,2,4,6],yticks=[-2,0,2,4,6])\n",
    "\n",
    "    plt.xlim((-2.9,7.5))\n",
    "    plt.ylim((-2.9,7.5))\n",
    "\n",
    "    plt.scatter(y=df[y], x=df[x], s=1, alpha=0.1,color='0.3', zorder=1)\n",
    "    plt.axhline(y=0, color='lightgrey', linestyle='dotted', zorder=0)\n",
    "    plt.axvline(x=0, color='lightgrey', linestyle='dotted', zorder=0)\n",
    "\n",
    "    ## 45 degree line\n",
    "    x2 = np.arange(-10,10,0.1)\n",
    "    y2 = x2\n",
    "    plt.plot(x2,y2, color='coral', alpha=0.8, zorder=4, linestyle='dotted')\n",
    "\n",
    "    plt.ylabel('Expression log2(fold change)', color='0.2')\n",
    "    plt.xlabel('log additive model', color='0.2')\n",
    "\n",
    "    k= '0.4'\n",
    "    ax.spines['right'].set_color(k)\n",
    "    ax.spines['left'].set_color(k)\n",
    "    ax.spines['top'].set_color(k)\n",
    "    ax.spines['bottom'].set_color(k)\n",
    "    ax.tick_params(axis='x', colors=k)\n",
    "    ax.tick_params(axis='y', colors=k)\n",
    "\n",
    "    plt.savefig('Fig1D.linmodel_%s.png' % key)\n",
    "\n",
    "\n",
    "with open(\"Fig1D.linmodel_score.txt\", \"w\") as results:\n",
    "    results.write('SEscore: %s\\n' % SEscore)\n",
    "    results.write('%s\\n' % SEfunc)\n",
    "    results.write('ESscore: %s\\n' % ESscore)\n",
    "    results.write('%s\\n' % ESfunc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5042d03",
   "metadata": {},
   "outputs": [],
   "source": [
    "### plotting Fig 1E\n",
    "\n",
    "df = pd.read_table(table)\n",
    "\n",
    "#### filtering out negative controls with significance\n",
    "df = df[~df['S'].str.contains('random') | (df['S'].str.contains('random') & (df['padj_S_run1'] > 0.001))]\n",
    "df = df.dropna(subset=['log2FC_ES', 'log2FC_SE'])\n",
    "\n",
    "annoA = {'2.REST','3.CTCF_TAD','4.CTCF_27ac', '5.GFI1', '6.YY1','7.active'}\n",
    "\n",
    "## add annotation of Silencers\n",
    "def dscri(x):\n",
    "    if 'CTCF_TAD' in x:\n",
    "        return '3.CTCF_TAD'\n",
    "    elif 'CTCF_chr' in  x:\n",
    "        return '4.CTCF_27ac'\n",
    "    elif 'pos' in x:\n",
    "        return '7.active'\n",
    "    elif 'random' in x:\n",
    "        return '1.random'\n",
    "    elif 'YY1' in x:\n",
    "        return '6.YY1'\n",
    "    elif 'REST' in x:\n",
    "        return '2.REST'\n",
    "    elif 'GFI1' in x:\n",
    "        return '5.GFI1'\n",
    "    else:\n",
    "        return '8.Blocker'\n",
    "\n",
    "disc = lambda x: dscri(x)\n",
    "df['annotation_S']= df['S'].apply(disc)\n",
    "df = df[~(df['annotation_S']=='8.Blocker')]\n",
    "\n",
    "df = df.sort_values(['annotation_S'], ascending=True)\n",
    "low = df[df['E'] == 'En02']\n",
    "high = df[df['E'] == 'En19']\n",
    "\n",
    "\n",
    "for library in ['ES', 'SE']:\n",
    "    key = 'log2FC_%s' % library\n",
    "\n",
    "    ### calc median of background controls\n",
    "    rndmL = low[(low['annotation_S']=='1.random')]\n",
    "    rndmL_med = pd.Series(rndmL[key]).median()\n",
    "\n",
    "    rndmH = high[(high['annotation_S']=='1.random')]\n",
    "    rndmH_med = pd.Series(rndmH[key]).median()\n",
    "\n",
    "    ## plot\n",
    "    fig = plt.figure(figsize=(3,2.5), dpi=300)\n",
    "\n",
    "    ymin, ymax = -4.5, 8.5\n",
    "    pl = sns.light_palette(\"seagreen\", n_colors=8)\n",
    "\n",
    "    ax = fig.add_subplot(121, yticks=[-2,0,2,4,6])\n",
    "    ax.set_aspect(1.4)\n",
    "    fig.subplots_adjust(left=0.2, bottom=0.2)\n",
    "    plt.ylim([-2,7])\n",
    "    plt.hlines([rndmL_med], ymin, ymax, \"coral\", linewidth=1, zorder=0,alpha=0.5)\n",
    "    sns.boxplot(x=low['annotation_S'],y=low[key], zorder=2, fliersize=0,palette=pl)\n",
    "    plt.ylabel('expression log2 (fold change)', color='0.2',fontsize='xx-small')\n",
    "    plt.xlabel('', color='0.2',fontsize='xx-small')\n",
    "    plt.xticks(rotation=40,ha='right',fontsize=\"xx-small\")\n",
    "    plt.yticks(fontsize='xx-small')\n",
    "    plt.text(0.5,6,'En02', color='0.2',fontsize='xx-small')\n",
    "\n",
    "    k='0.4'\n",
    "    ax.spines['right'].set_color(k)\n",
    "    ax.spines['left'].set_color(k)\n",
    "    ax.spines['top'].set_color(k)\n",
    "    ax.spines['bottom'].set_color(k)\n",
    "    ax.tick_params(axis='x', colors=k)\n",
    "    ax.tick_params(axis='y', colors=k)\n",
    "\n",
    "\n",
    "    ax2 = fig.add_subplot(122, yticks=[-2,0,2,4,6])\n",
    "    ax2.set_aspect(1.4)\n",
    "    fig.subplots_adjust(left=0.2, bottom=0.2)\n",
    "    plt.ylim([-2,7])\n",
    "    plt.hlines([rndmH_med], ymin, ymax, \"coral\",linewidth=1, zorder=0,alpha=0.5)\n",
    "    sns.boxplot(x=high['annotation_S'],y=high[key], zorder=2, fliersize=0,palette=pl)\n",
    "    plt.ylabel('', color='0.2',fontsize='xx-small')\n",
    "    plt.xlabel('', color='0.2',fontsize='xx-small')\n",
    "    plt.xticks(rotation=40,ha='right',fontsize=\"xx-small\")\n",
    "    plt.yticks(fontsize='xx-small')\n",
    "    plt.text(0.5,6,'En19', color='0.2',fontsize='xx-small')\n",
    "\n",
    "\n",
    "    k='0.4'\n",
    "    ax2.spines['right'].set_color(k)\n",
    "    ax2.spines['left'].set_color(k)\n",
    "    ax2.spines['top'].set_color(k)\n",
    "    ax2.spines['bottom'].set_color(k)\n",
    "    ax2.tick_params(axis='x', colors=k)\n",
    "    ax2.tick_params(axis='y', colors=k)\n",
    "\n",
    "    fig.savefig('Fig1E.%s.box.pdf' % library, transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "356e54ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "### plotting Fig 1F\n",
    "\n",
    "df = pd.read_table(table)\n",
    "\n",
    "#### filtering out negative controls with significance\n",
    "df = df[~df['S'].str.contains('random') | (df['S'].str.contains('random') & (df['padj_S_run1'] > 0.001))]\n",
    "df = df.dropna(subset=['log2FC_ES', 'log2FC_SE'])\n",
    "\n",
    "#annoA = {'1.CTCF_nonTAD','2.CTCF_TAD','3.GFI1','4.REST', '5.YY1', '6.active'}\n",
    "\n",
    "### P's rank is based on the P's activity on run2\n",
    "dfp = df.sort_values(by='E',ascending=True)\n",
    "dfp = pd.Series(dfp['E'])\n",
    "dfp = dfp.drop_duplicates()\n",
    "lp = dfp.values.tolist()\n",
    "\n",
    "### make an empty dataframe\n",
    "output = pd.DataFrame(columns=['S', 'E', 'log2FC_E_run1', 'log2FC_E_run2','zES', 'zSE'])\n",
    "\n",
    "### calc z scores\n",
    "\n",
    "for p in lp:\n",
    "    df1 = df[df['E'] == p].reset_index(drop=True)\n",
    "    rndm = df1[df1['annotation_S'] == '7.random_genomic']\n",
    "    rndm = rndm.dropna(subset=['log2FC_ES'])\n",
    "\n",
    "    if len(rndm) >0:\n",
    "        df1['zES'] = (df1['log2FC_ES'] - rndm['log2FC_ES'].mean()) / rndm['log2FC_ES'].std(ddof=0)\n",
    "        df1['zSE'] = (df1['log2FC_SE'] - rndm['log2FC_SE'].mean()) / rndm['log2FC_SE'].std(ddof=0)\n",
    "\n",
    "        df1 = df1.loc[:,['S','E','log2FC_E_run1', 'log2FC_E_run2', 'zES', 'zSE']]\n",
    "\n",
    "        output = pd.concat([output, df1], axis=0)\n",
    "\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "\n",
    "key = 'REST'\n",
    "\n",
    "output = output[output['S'].str.contains(key)]\n",
    "output = output.dropna(subset=['zES'])\n",
    "\n",
    "xmin, xmax = -20,20\n",
    "fig = plt.figure(figsize=(16,16))\n",
    "\n",
    "plt.subplot(2,1,1)\n",
    "sns.boxplot(data = output, x='E', y='zES', color=\"darkgrey\",fliersize=0)\n",
    "plt.xticks(rotation=30)\n",
    "plt.ylim(-4.5,4.5)\n",
    "plt.hlines([0],xmin, xmax, \"0.7\", linestyles='dashed', zorder=0)\n",
    "plt.ylabel('zscore ES')\n",
    "\n",
    "plt.subplot(2,1,2)\n",
    "sns.boxplot(data = output, x='E', y='zSE', color=\"darkgrey\",fliersize=0)\n",
    "plt.xticks(rotation=30)\n",
    "plt.ylim(-5,5)\n",
    "plt.hlines([0],xmin, xmax, \"0.7\", linestyles='dashed', zorder=0)\n",
    "plt.ylabel('zscore SE')\n",
    "\n",
    "plt.savefig('Fig1F_box.REST.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e14d1ee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### plotting Fig 3B\n",
    "\n",
    "file = 'processed_data/RESTscreen_derived_K562_results.run1.txt'\n",
    "l2fc = pd.read_table(file)\n",
    "\n",
    "## select En19\n",
    "l2fc = l2fc[l2fc['ID'].str.contains('En19')]\n",
    "\n",
    "## separate results to canonical, negative control, and no-canonical\n",
    "ref = l2fc[l2fc['ID'].str.contains('_Ref')]\n",
    "Neg = l2fc[l2fc['project'] == 'NegCtrl']\n",
    "l2fc = l2fc[l2fc['project'] == 'noMotif']\n",
    "\n",
    "### process canonical motifs\n",
    "#### make a dictionary to call score\n",
    "refsc = pd.read_table('processed_data/ref_all_fimo.txt', sep='\\t')\n",
    "refsc = refsc.set_index('sequence name')\n",
    "refsc_dic = refsc['score'].to_dict()\n",
    "def Refscore(x):\n",
    "    return refsc_dic.get(x,0)\n",
    "\n",
    "#### select RE1 with strong canonical motif\n",
    "idref = ref['ID'].str.split('^',expand=True)\n",
    "idref.columns = ['enhancer','silencer']\n",
    "ref = pd.concat([ref,idref],axis=1)\n",
    "ref = ref.loc[:,['enhancer','silencer','log2FoldChange']].reset_index(drop=True)\n",
    "ref['score'] = ref['silencer'].apply(Refscore)\n",
    "ref = ref[ref['score'] > 20.86].dropna(subset= ['log2FoldChange'])\n",
    "\n",
    "#### process nomotifs\n",
    "ids = l2fc['ID'].str.split('^',expand=True)\n",
    "ids.columns = ['enhancer','silencer']\n",
    "l2fc = pd.concat([l2fc,ids],axis=1)\n",
    "l2fc = l2fc.loc[:,['enhancer','silencer','log2FoldChange']].reset_index(drop=True)\n",
    "\n",
    "\n",
    "## FIMO result for left and right half motif\n",
    "left = pd.read_table('processed_data/halfmotif_left_fimo.txt')\n",
    "right = pd.read_table('processed_data/halfmotif_right_fimo.txt')\n",
    "\n",
    "### make dictionaries to call half-score\n",
    "lsc = left.loc[:,['sequence_name','score']].sort_values('score',ascending=False)\n",
    "lsc = lsc.drop_duplicates(subset=['sequence_name'])\n",
    "lsc = lsc.set_index('sequence_name')\n",
    "dicL = lsc['score'].to_dict()\n",
    "def Lscore(x):\n",
    "    return dicL.get(x,0)\n",
    "\n",
    "rsc = right.loc[:,['sequence_name','score']].sort_values('score',ascending=False)\n",
    "rsc = rsc.drop_duplicates(subset=['sequence_name'])\n",
    "rsc = rsc.set_index('sequence_name')\n",
    "dicR = rsc['score'].to_dict()\n",
    "def Rscore(x):\n",
    "    return dicR.get(x,0)\n",
    "\n",
    "l2fc['score_L'] = l2fc['silencer'].apply(Lscore)\n",
    "l2fc['score_R'] = l2fc['silencer'].apply(Rscore)\n",
    "\n",
    "### make a table for ones with both L and R halfs\n",
    "both = list(set(left['sequence_name'].values.tolist())  & set(right['sequence_name'].values.tolist()))\n",
    "\n",
    "cmn_l = left[left['sequence_name'].isin(both)]\n",
    "cmn_l = cmn_l.drop_duplicates(subset=['sequence_name']).sort_values(by='sequence_name').set_index('sequence_name')\n",
    "cmn_l = cmn_l.drop(['motif_id','motif_alt_id','q-value'],axis=1)\n",
    "cmn_l.columns = ['startL','stopL','strandL','scoreL','p-valueL','mat_seq_L']\n",
    "\n",
    "cmn_r = right[right['sequence_name'].isin(both)]\n",
    "cmn_r = cmn_r.drop_duplicates(subset=['sequence_name']).sort_values(by='sequence_name').set_index('sequence_name')\n",
    "cmn_r = cmn_r.drop(['motif_id','motif_alt_id','q-value'],axis=1)\n",
    "cmn_r.columns = ['startR','stopR','strandR','scoreR','p-valueR','mat_seq_R']\n",
    "\n",
    "df = pd.concat([cmn_l,cmn_r],axis=1)\n",
    "\n",
    "### filter weak motifs\n",
    "df['sum_score'] = df['scoreL'] + df['scoreR']\n",
    "df = df[df['sum_score']>=20.86]\n",
    "\n",
    "\n",
    "df = df.reset_index()\n",
    "df['gap'] = 'NA'\n",
    "df['alignment'] = 'NA'\n",
    "df['strands'] = df['strandL'] + df['strandR']\n",
    "\n",
    "\n",
    "## categorize non-canonical motifs\n",
    "for i in range(len(df)):\n",
    "    strands = df.loc[i,'strands']\n",
    "\n",
    "    if df.loc[i,'stopL'] < df.loc[i,'startR']:\n",
    "        gap = df.loc[i,'startR'] - df.loc[i,'stopL'] -1\n",
    "        if  strands == '++':\n",
    "            alignment = 'atypically_spaced'\n",
    "        elif strands == '--' :\n",
    "            alignment = 'flipped'\n",
    "        elif strands == '+-':\n",
    "            alignment = 'convergent'\n",
    "        else:\n",
    "            alignment = 'divergent'\n",
    "\n",
    "    elif df.loc[i,'stopR'] < df.loc[i,'startL']:\n",
    "        gap = df.loc[i,'startL'] - df.loc[i,'stopR'] -1\n",
    "        if strands == '++':\n",
    "            alignment = 'flipped'\n",
    "        elif strands == '--' :\n",
    "            alignment = 'atypically_spaced'\n",
    "        elif strands == '+-':\n",
    "            alignment = 'convergent'\n",
    "        else:\n",
    "            alignment = 'divergent'\n",
    "\n",
    "    else:\n",
    "        gap = 'NA'\n",
    "        alignment = 'overlap'\n",
    "\n",
    "    df.loc[i,'gap'] = gap\n",
    "    df.loc[i,'alignment'] = alignment\n",
    "\n",
    "\n",
    "df = df[(df['gap']>=0)&(df['gap']<200)]\n",
    "df = df.astype({'gap': int})\n",
    "\n",
    "df = df.set_index('sequence_name')\n",
    "\n",
    "\n",
    "dic_gap = df['gap'].to_dict()\n",
    "def gap(x):\n",
    "    return dic_gap.get(x,'NA')\n",
    "\n",
    "dic_algn = df['alignment'].to_dict()\n",
    "def algn(x):\n",
    "    return dic_algn.get(x,'others')\n",
    "\n",
    "l2fc['gap'] = l2fc['silencer'].apply(gap)\n",
    "l2fc['alignment'] = l2fc['silencer'].apply(algn)\n",
    "\n",
    "\n",
    "for i in range(len(l2fc)):\n",
    "    algn = l2fc.loc[i,'alignment']\n",
    "    if algn == 'others':\n",
    "        if l2fc.loc[i,'score_L'] > 10:\n",
    "            g = 'left_only'\n",
    "        elif l2fc.loc[i,'score_R'] > 10:\n",
    "            g = 'right_only'\n",
    "        else:\n",
    "            g = 'No_half'\n",
    "        l2fc.loc[i,'alignment'] = g\n",
    "        l2fc.loc[i,'gap'] = 0\n",
    "\n",
    "sortl = ['canonical','No_half','atypically_spaced','flipped','convergent','divergent','left_only','right_only']\n",
    "l2fc['sort'] = l2fc['alignment'].apply(lambda x: sortl.index(x) if x in sortl else -1)\n",
    "l2fc = l2fc.sort_values(by='sort')\n",
    "\n",
    "l2fc = l2fc[l2fc['gap'] <= 100]\n",
    "\n",
    "\n",
    "## add negative control\n",
    "Neg['score_L'] = 0\n",
    "Neg['score_R'] = 0\n",
    "Neg['gap'] = 'NA'\n",
    "Neg['alignment'] = 'NegCtrl'\n",
    "negmed = Neg['log2FoldChange'].median()\n",
    "\n",
    "## add canonical\n",
    "ref['score_L'] = 0\n",
    "ref['score_R'] = 0\n",
    "ref['gap'] = 'NA'\n",
    "ref['alignment'] = 'canonical'\n",
    "\n",
    "l2fc = pd.concat([Neg, ref, l2fc],axis=0)\n",
    "\n",
    "\n",
    "## plot\n",
    "fig1 = plt.figure(figsize=(2,2), dpi=300)\n",
    "fig1 = plt.xticks(fontsize= 'xx-small',rotation = 20)\n",
    "fig1 = plt.yticks(fontsize= 'xx-small')\n",
    "fig1 = plt.ylim((-3,8.5))\n",
    "fig1 = plt.ylabel('log2FoldChange',fontsize='xx-small')\n",
    "fig1 =  sns.boxplot(y=l2fc['log2FoldChange'], x=l2fc['alignment'],linewidth=1,zorder=1,showfliers = False)\n",
    "fig1 = plt.axhline(y=negmed, color='grey', alpha=0.5,linewidth=1,zorder=0)\n",
    "fig1 = plt.title('nomotifs expression (K562,En19)',fontsize='xx-small')\n",
    "fig1 = plt.savefig('Fig3B_box.pdf')\n",
    "\n",
    "\n",
    "## calc p and adjp values\n",
    "p = []\n",
    "\n",
    "for i in sortl:\n",
    "    l2fc_nc = l2fc[l2fc['alignment'] == 'NegCtrl']\n",
    "    l2fc_pd = l2fc[l2fc['alignment'] == i]\n",
    "\n",
    "    pval = stats.mannwhitneyu(l2fc_nc['log2FoldChange'].dropna(),l2fc_pd['log2FoldChange'].dropna() , alternative='two-sided',use_continuity=True).pvalue\n",
    "    p.append(pval)\n",
    "    #print i,'\\t',pval\n",
    "\n",
    "\n",
    "adjp = multi.multipletests(np.array(p),alpha=0.05,method='fdr_bh')\n",
    "adjp = adjp[1].tolist()\n",
    "res = pd.Series(adjp,index=sortl)\n",
    "\n",
    "res.to_csv('Fig3B_stat.txt',index=True,header=False,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9488a431",
   "metadata": {},
   "outputs": [],
   "source": [
    "### plotting Fig 3C\n",
    "\n",
    "### import files\n",
    "left = pd.read_table('processed_data/halfmotif_left_fimo.txt')\n",
    "right = pd.read_table('processed_data/halfmotif_right_fimo.txt')\n",
    "\n",
    "\n",
    "### make dataframe for ones with both L and R halfs\n",
    "both = list(set(left['sequence_name'].values.tolist())  & set(right['sequence_name'].values.tolist()))\n",
    "\n",
    "cmn_l = left[left['sequence_name'].isin(both)]\n",
    "cmn_l = cmn_l.drop_duplicates(subset=['sequence_name']).sort_values(by='sequence_name').set_index('sequence_name')\n",
    "cmn_l = cmn_l.drop(['motif_id','motif_alt_id','q-value'],axis=1)\n",
    "cmn_l.columns = ['startL','stopL','strandL','scoreL','p-valueL','mat_seq_L']\n",
    "\n",
    "\n",
    "cmn_r = right[right['sequence_name'].isin(both)]\n",
    "cmn_r = cmn_r.drop_duplicates(subset=['sequence_name']).sort_values(by='sequence_name').set_index('sequence_name')\n",
    "cmn_r = cmn_r.drop(['motif_id','motif_alt_id','q-value'],axis=1)\n",
    "cmn_r.columns = ['startR','stopR','strandR','scoreR','p-valueR','mat_seq_R']\n",
    "\n",
    "df = pd.concat([cmn_l,cmn_r],axis=1)\n",
    "\n",
    "\n",
    "### cut off weak motifs\n",
    "df['sum_score'] = df['scoreL'] + df['scoreR']\n",
    "df = df[df['sum_score']>=20.86]\n",
    "\n",
    "df = df.reset_index()\n",
    "df['gap'] = 'NA'\n",
    "df['alignment'] = 'NA'\n",
    "df['strands'] = df['strandL'] + df['strandR']\n",
    "\n",
    "## categorize non-canonical motifs\n",
    "for i in range(len(df)):\n",
    "    strands = df.loc[i,'strands']\n",
    "\n",
    "    if df.loc[i,'stopL'] < df.loc[i,'startR']:\n",
    "        gap = df.loc[i,'startR'] - df.loc[i,'stopL'] -1\n",
    "        if  strands == '++':\n",
    "            alignment = 'atypically_spaced'\n",
    "        elif strands == '--' :\n",
    "            alignment = 'flipped'\n",
    "        elif strands == '+-':\n",
    "            alignment = 'convergent'\n",
    "        else:\n",
    "            alignment = 'divergent'\n",
    "\n",
    "    elif df.loc[i,'stopR'] < df.loc[i,'startL']:\n",
    "        gap = df.loc[i,'startL'] - df.loc[i,'stopR'] -1\n",
    "        if strands == '++':\n",
    "            alignment = 'flipped'\n",
    "        elif strands == '--' :\n",
    "            alignment = 'atypically_spaced'\n",
    "        elif strands == '+-':\n",
    "            alignment = 'convergent'\n",
    "        else:\n",
    "            alignment = 'divergent'\n",
    "\n",
    "    else:\n",
    "        gap = 'NA'\n",
    "        alignment = 'overlap'\n",
    "\n",
    "    df.loc[i,'gap'] = gap\n",
    "    df.loc[i,'alignment'] = alignment\n",
    "\n",
    "df = df[(df['gap']>=0)&(df['gap']<=25)]\n",
    "df = df.astype({'gap': int})\n",
    "\n",
    "\n",
    "## plot\n",
    "bins = np.arange(0, 25, 1)\n",
    "t = 60\n",
    "df = df[df['alignment'] != 'overlap']\n",
    "LR = df[df['alignment'] == 'atypically_spaced']\n",
    "LR = LR['gap'].values\n",
    "RL = df[df['alignment'] == 'flipped']\n",
    "RL = RL['gap'].values\n",
    "inw = df[df['alignment'] == 'convergent']\n",
    "inw = inw['gap'].values\n",
    "outw = df[df['alignment'] == 'divergent']\n",
    "outw = outw['gap'].values\n",
    "\n",
    "fig1 = plt.figure(figsize=(4,4), dpi=300)\n",
    "fig1 = plt.title('gap (bp) between half motifs', color='0.2')\n",
    "\n",
    "ax1 = plt.subplot(221)\n",
    "ax1 = plt.title('atypically_spaced',fontsize='xx-small')\n",
    "ax1 = plt.ylim((0,t))\n",
    "ax1 = plt.xlim((0,25))\n",
    "ax1 = plt.hist(LR,bins)\n",
    "ax1 = plt.xticks([0,25,50],fontsize='xx-small')\n",
    "ax1 = plt.yticks(fontsize='xx-small')\n",
    "\n",
    "ax2 = plt.subplot(222)\n",
    "ax2 = plt.title('flipped',fontsize='xx-small')\n",
    "ax2 = plt.ylim((0,t))\n",
    "ax2 = plt.xlim((0,25))\n",
    "ax2 = plt.hist(RL,bins)\n",
    "ax2 = plt.xticks([0,25,50],fontsize='xx-small')\n",
    "ax2 = plt.yticks(fontsize='xx-small')\n",
    "\n",
    "ax3 = plt.subplot(223)\n",
    "ax3 = plt.title('convergent',fontsize='xx-small')\n",
    "ax3 = plt.ylim((0,t))\n",
    "ax3 = plt.xlim((0,25))\n",
    "ax3 = plt.hist(inw,bins)\n",
    "ax3 = plt.xticks([0,25,50],fontsize='xx-small')\n",
    "ax3 = plt.yticks(fontsize='xx-small')\n",
    "\n",
    "ax4 = plt.subplot(224)\n",
    "ax4 = plt.title('divergent',fontsize='xx-small')\n",
    "ax4 = plt.ylim((0,t))\n",
    "ax4 = plt.xlim((0,25))\n",
    "ax4 = plt.hist(outw,bins)\n",
    "ax4 = plt.xticks([0,25,50],fontsize='xx-small')\n",
    "ax4 = plt.yticks(fontsize='xx-small')\n",
    "\n",
    "fig1 = plt.savefig('Fig3C_hist.pdf' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5eac751",
   "metadata": {},
   "outputs": [],
   "source": [
    "### plotting Fig 3D\n",
    "\n",
    "l2fc = pd.read_table('processed_data/RESTscreen_derived_K562_results.run1.txt')\n",
    "\n",
    "### select negative controls and no-canonicals\n",
    "l2fc = l2fc[l2fc['ID'].str.contains('En19')]\n",
    "Neg = l2fc[l2fc['project'] == 'NegCtrl']\n",
    "l2fc = l2fc[l2fc['project'] == 'noMotif']\n",
    "\n",
    "\n",
    "ids = l2fc['ID'].str.split('^',expand=True)\n",
    "ids.columns = ['enhancer','silencer']\n",
    "l2fc = pd.concat([l2fc,ids],axis=1)\n",
    "l2fc = l2fc.loc[:,['enhancer','silencer','log2FoldChange']].reset_index(drop=True)\n",
    "\n",
    "### inport half-scores\n",
    "left = pd.read_table('processed_data/halfmotif_left_fimo.txt')\n",
    "right = pd.read_table('processed_data/halfmotif_right_fimo.txt')\n",
    "\n",
    "#### make dictionaries to call half-scores\n",
    "lsc = left.loc[:,['sequence_name','score']].sort_values('score',ascending=False)\n",
    "lsc = lsc.drop_duplicates(subset=['sequence_name'])\n",
    "lsc = lsc.set_index('sequence_name')\n",
    "dicL = lsc['score'].to_dict()\n",
    "def Lscore(x):\n",
    "    return dicL.get(x,0)\n",
    "\n",
    "rsc = right.loc[:,['sequence_name','score']].sort_values('score',ascending=False)\n",
    "rsc = rsc.drop_duplicates(subset=['sequence_name'])\n",
    "rsc = rsc.set_index('sequence_name')\n",
    "dicR = rsc['score'].to_dict()\n",
    "def Rscore(x):\n",
    "    return dicR.get(x,0)\n",
    "\n",
    "l2fc['score_L'] = l2fc['silencer'].apply(Lscore)\n",
    "l2fc['score_R'] = l2fc['silencer'].apply(Rscore)\n",
    "\n",
    "### make a dataframe for ones with both L and R halfs\n",
    "both = list(set(left['sequence_name'].values.tolist())  & set(right['sequence_name'].values.tolist()))\n",
    "\n",
    "cmn_l = left[left['sequence_name'].isin(both)]\n",
    "cmn_l = cmn_l.drop_duplicates(subset=['sequence_name']).sort_values(by='sequence_name').set_index('sequence_name')\n",
    "cmn_l = cmn_l.drop(['motif_id','motif_alt_id','q-value'],axis=1)\n",
    "cmn_l.columns = ['startL','stopL','strandL','scoreL','p-valueL','mat_seq_L']\n",
    "\n",
    "cmn_r = right[right['sequence_name'].isin(both)]\n",
    "cmn_r = cmn_r.drop_duplicates(subset=['sequence_name']).sort_values(by='sequence_name').set_index('sequence_name')\n",
    "cmn_r = cmn_r.drop(['motif_id','motif_alt_id','q-value'],axis=1)\n",
    "cmn_r.columns = ['startR','stopR','strandR','scoreR','p-valueR','mat_seq_R']\n",
    "\n",
    "df = pd.concat([cmn_l,cmn_r],axis=1)\n",
    "\n",
    "### cut off weak motifs\n",
    "df['sum_score'] = df['scoreL'] + df['scoreR']\n",
    "df = df[df['sum_score']>=20.86]\n",
    "\n",
    "df = df.reset_index()\n",
    "df['gap'] = 'NA'\n",
    "df['alignment'] = 'NA'\n",
    "df['strands'] = df['strandL'] + df['strandR']\n",
    "\n",
    "## categorize non-canonical motifs\n",
    "for i in range(len(df)):\n",
    "    strands = df.loc[i,'strands']\n",
    "\n",
    "    if df.loc[i,'stopL'] < df.loc[i,'startR']:\n",
    "        gap = df.loc[i,'startR'] - df.loc[i,'stopL'] -1\n",
    "        if  strands == '++':\n",
    "            alignment = 'atypically_spaced'\n",
    "        elif strands == '--' :\n",
    "            alignment = 'flipped'\n",
    "        elif strands == '+-':\n",
    "            alignment = 'convergent'\n",
    "        else:\n",
    "            alignment = 'divergent'\n",
    "\n",
    "    elif df.loc[i,'stopR'] < df.loc[i,'startL']:\n",
    "        gap = df.loc[i,'startL'] - df.loc[i,'stopR'] -1\n",
    "        if strands == '++':\n",
    "            alignment = 'flipped'\n",
    "        elif strands == '--' :\n",
    "            alignment = 'atypically_spaced'\n",
    "        elif strands == '+-':\n",
    "            alignment = 'divergent'\n",
    "        else:\n",
    "            alignment = 'convergent'\n",
    "\n",
    "    else:\n",
    "        gap = 'NA'\n",
    "        alignment = 'overlap'\n",
    "\n",
    "    df.loc[i,'gap'] = gap\n",
    "    df.loc[i,'alignment'] = alignment\n",
    "\n",
    "df = df[(df['gap']>=0)&(df['gap']<200)]\n",
    "df = df.astype({'gap': int})\n",
    "\n",
    "df = df.set_index('sequence_name')\n",
    "\n",
    "\n",
    "dic_gap = df['gap'].to_dict()\n",
    "def gap(x):\n",
    "    return dic_gap.get(x,'NA')\n",
    "\n",
    "dic_algn = df['alignment'].to_dict()\n",
    "def algn(x):\n",
    "    return dic_algn.get(x,'others')\n",
    "\n",
    "l2fc['gap'] = l2fc['silencer'].apply(gap)\n",
    "l2fc['alignment'] = l2fc['silencer'].apply(algn)\n",
    "\n",
    "## select atypically spaced strong non-canonical motifs with <25 bp gap\n",
    "l2fc = l2fc[l2fc['alignment'] == 'atypically_spaced']\n",
    "l2fc = l2fc[l2fc['gap'] <= 25]\n",
    "l2fc = l2fc.sort_values(by='gap').reset_index(drop=True)\n",
    "l2fc = l2fc[(l2fc['score_L'] + l2fc['score_R']) > 20.86]\n",
    "\n",
    "## add dummy data to plot gap=5 and gap=6\n",
    "l2fc = l2fc.loc[:,['ID','alignment','gap','score_L','score_R','log2FoldChange']]\n",
    "gap5 = pd.Series(['dummy','LR',5,0,0,7], index=l2fc.columns)\n",
    "gap6 = pd.Series(['dummy','LR',6,0,0,7], index=l2fc.columns)\n",
    "l2fc =l2fc.append(gap5, ignore_index=True)\n",
    "l2fc =l2fc.append(gap6, ignore_index=True)\n",
    "\n",
    "l2fc = l2fc.sort_values(by='gap')\n",
    "\n",
    "## concatenate long gap motif\n",
    "for i in range(len(l2fc)):\n",
    "    if l2fc.loc[i,'gap'] >= 12:\n",
    "        l2fc.loc[i,'gap'] = '>=12'\n",
    "\n",
    "## add negative control\n",
    "Neg['score_L'] = 0\n",
    "Neg['score_R'] = 0\n",
    "Neg['gap'] = 'Neg'\n",
    "Neg['alignment'] = 'NegCtrl'\n",
    "negmed = Neg['log2FoldChange'].median()\n",
    "Neg = Neg.loc[:,['ID','alignment','gap','score_L','score_R','log2FoldChange']]\n",
    "\n",
    "l2fc = pd.concat([Neg,l2fc],axis=0)\n",
    "\n",
    "l = ['Neg',0,1,2,3,4,5,6,7,8,9,10,11,'>=12']\n",
    "l2fc = l2fc[l2fc['gap'].isin(l)]\n",
    "\n",
    "## plot\n",
    "fig1 = plt.figure(figsize=(2.25,2), dpi=300)\n",
    "fig1 = plt.xlabel('')\n",
    "fig1 = plt.xticks(fontsize= 'xx-small',rotation = 30)\n",
    "fig1 = plt.yticks(fontsize= 'xx-small')\n",
    "fig1 = plt.ylim((-3,8.5))\n",
    "fig1 =  sns.boxplot(y=l2fc['log2FoldChange'], x=l2fc['gap'],linewidth=1,zorder=1,showfliers = False)\n",
    "fig1 = plt.axhline(y=negmed, color='grey', alpha=0.5,linewidth=1,zorder=0)\n",
    "fig1 = plt.title('nomotifs expression (K562,En19)')\n",
    "fig1 = plt.savefig('Fig3D_gapbox.pdf')\n",
    "\n",
    "\n",
    "## calc p and adj values\n",
    "Neg = Neg['log2FoldChange'].dropna().values\n",
    "p = []\n",
    "gap = []\n",
    "for i in range(26):\n",
    "    test = l2fc[l2fc['gap'] == i]\n",
    "    if len(test) > 3:\n",
    "        test = test['log2FoldChange'].dropna().values\n",
    "        pval = stats.mannwhitneyu(Neg,test , alternative='two-sided',use_continuity=True).pvalue\n",
    "        p.append(pval)\n",
    "        gap.append(i)\n",
    "\n",
    "adjp = multi.multipletests(np.array(p),alpha=0.05,method='fdr_bh')\n",
    "adjp = adjp[1].tolist()\n",
    "res = pd.Series(adjp,index=gap)\n",
    "\n",
    "res.to_csv('Fig3D_stat.txt',index=True,header=False,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c1720dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "### plotting Fig 5B\n",
    "\n",
    "seg = 20.86\n",
    "enhancer = 'En19'\n",
    "\n",
    "\n",
    "df = pd.read_table('processed_data/RESTscreen_AF_K562_emVAR.out', sep='\\t')\n",
    "df = df.dropna(subset=['LogSkew'])\n",
    "df = df[df['ID'].str.contains(enhancer)]\n",
    "df = df[df['ID'].str.contains(':wP')]\n",
    "\n",
    "\n",
    "### making a column for SNP position\n",
    "spl = df['ID'].str.split(':wP',expand=True)\n",
    "spl.columns = ['sil','SNP_pos']\n",
    "df = pd.concat([df,spl['SNP_pos']],axis=1)\n",
    "df['SNP_pos'] = df['SNP_pos'].astype(int)\n",
    "\n",
    "df = df.reset_index(drop=True)\n",
    "\n",
    "### calc delta binding score\n",
    "ref = pd.read_table('processed_data/variants_ref_fimo.txt', sep='\\t')\n",
    "ref = ref[(ref['start']== 91) | (ref['start']== 92)]\n",
    "ref = ref.sort_values(by='#pattern name', ascending=False) # prioritize MA0138.2\n",
    "ref = ref.drop_duplicates(subset=['sequence name'])\n",
    "spl_r = ref['sequence name'].str.split(':R:wP', expand=True)\n",
    "spl_r.columns = ['SNP', 'position']\n",
    "ref = pd.concat([ref, spl_r['SNP']],axis=1).set_index('SNP')\n",
    "ref_score = ref['score'].to_dict()\n",
    "ref_strand = ref['strand'].to_dict()\n",
    "\n",
    "alt = pd.read_table('processed_data/variants_alt_fimo.txt', sep='\\t')\n",
    "alt = alt[(alt['start']== 91) | (alt['start']== 92)]\n",
    "alt = alt.sort_values(by='#pattern name', ascending=False)\n",
    "alt = alt.drop_duplicates(subset=['sequence name'])\n",
    "spl_a = alt['sequence name'].str.split(':A:wP', expand=True)\n",
    "spl_a.columns = ['SNP', 'position']\n",
    "alt = pd.concat([alt, spl_a['SNP']],axis=1).set_index('SNP')\n",
    "alt_score = alt['score'].to_dict()\n",
    "alt_strand = alt['strand'].to_dict()\n",
    "\n",
    "\n",
    "df['score_maj'] = 0\n",
    "df['score_min'] = 0\n",
    "df['strand'] = 'NA'\n",
    "\n",
    "df = df.reset_index(drop=True)\n",
    "for i in range(len(df)):\n",
    "    id = df.loc[i,'ID']\n",
    "    s = df.loc[i,'SNP']\n",
    "    if 'R:wP' in id:\n",
    "        df.loc[i,'score_maj'] = ref_score.get(s,0)\n",
    "        df.loc[i,'score_min'] = alt_score.get(s,0)\n",
    "        df.loc[i,'strand'] = ref_strand.get(s,0)\n",
    "\n",
    "\n",
    "    elif 'A:wP' in id:\n",
    "        df.loc[i,'score_maj'] = alt_score.get(s,0)\n",
    "        df.loc[i,'score_min'] = ref_score.get(s,0)\n",
    "        df.loc[i,'strand'] = alt_strand.get(s,0)\n",
    "\n",
    "\n",
    "df = df[(df['score_maj'] > 0) & (df['score_min'] > 0)]\n",
    "df['dBS'] = df['score_min'] - df['score_maj']\n",
    "\n",
    "### orient motifs in a same direction\n",
    "df['rel_pos'] = 0\n",
    "df = df.reset_index(drop=True)\n",
    "for i in range(len(df)):\n",
    "    pos = df.loc[i,'SNP_pos']\n",
    "    str = df.loc[i,'strand']\n",
    "    if str == '+':\n",
    "        df.loc[i,'rel_pos'] = pos\n",
    "    elif str == '-':\n",
    "        df.loc[i,'rel_pos'] = 202 - pos\n",
    "\n",
    "\n",
    "### taking absolute value of logSkew\n",
    "df['abs_skew'] = df['LogSkew'].abs()\n",
    "\n",
    "df['dist'] = abs(101 - df['rel_pos'])\n",
    "df = df.sort_values(by='dist')\n",
    "df = df.drop_duplicates(subset=['SNP'])\n",
    "\n",
    "df = df[(df['rel_pos'] <= 126) & (df['rel_pos'] >= 76)]\n",
    "\n",
    "\n",
    "### separate above and below for line plots\n",
    "above = df[(df['score_min'] > seg)| (df['score_maj'] > seg)]\n",
    "below = df[(df['score_min'] < seg) & (df['score_maj'] < seg)]\n",
    "\n",
    "\n",
    "\n",
    "### scatter: position-vs-skew\n",
    "fig2 = plt.figure(figsize=(3,3), dpi=300)\n",
    "boxprops = dict(linewidth=0.5, color='0.6')\n",
    "\n",
    "ax2 = plt.subplot(211)\n",
    "ax2.set(ylim=(-0.1, 3.5))\n",
    "ax2.set(xlim=(76, 126))\n",
    "ax2 = sns.boxplot(y=above['LogSkew'].abs(), x=above['rel_pos'],fliersize=0,color='coral',boxprops=dict(linewidth=1))\n",
    "ax2 = plt.ylabel('abs_LogSkew_above', color='0.2', fontsize=6)\n",
    "ax2 = plt.xlabel('')\n",
    "plt.xticks(rotation=90,fontsize=4)\n",
    "plt.yticks(fontsize=5)\n",
    "\n",
    "\n",
    "ax3 = plt.subplot(212)\n",
    "ax3.set(ylim=(-0.1, 3.5))\n",
    "ax3.set(xlim=(76, 126))\n",
    "ax3 = sns.boxplot(y=below['LogSkew'].abs(), x=below['rel_pos'],fliersize=0,color=\"darkgrey\")\n",
    "ax3 = plt.ylabel('abs_LogSkew_below', color='0.2',fontsize=6)\n",
    "ax3 = plt.xlabel('SNP position (relative)', color='0.2', fontsize=4)\n",
    "plt.xticks(rotation=90, fontsize=4)\n",
    "plt.yticks(fontsize=5)\n",
    "\n",
    "fig2 =  plt.savefig('Fig5b_abs_skew_box.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2102432",
   "metadata": {},
   "outputs": [],
   "source": [
    "### plotting Fig 5C\n",
    "\n",
    "### select one result for one SNP (select the closest one to the center)\n",
    "df['distance'] = abs(101 - df['SNP_pos'])\n",
    "df = df.sort_values(by = 'distance')\n",
    "df = df.drop_duplicates(subset=['SNP'])\n",
    "\n",
    "\n",
    "### select \"inside\" SNP\n",
    "df = df[(df['SNP_pos']>=91) & (df['SNP_pos']<=111)]\n",
    "\n",
    "df =df.dropna(subset=['LogSkew'])\n",
    "\n",
    "\n",
    "### remove outliers of dBS derived from mis score calc\n",
    "df = df[(df['dBS'] > -20) & (df['dBS'] < 20)]\n",
    "\n",
    "### plots\n",
    "for segment in ['above', 'below']:\n",
    "\n",
    "    if segment == 'above':\n",
    "        df1 = df[(df['score_maj'] > seg)| (df['score_min'] > seg)]\n",
    "    elif segment == 'below':\n",
    "        df1 = df[(df['score_maj'] < seg) & (df['score_min'] < seg)]\n",
    "\n",
    "    skew = np.array(df1['LogSkew'].tolist()).reshape(-1,1)\n",
    "    dbs = np.array(df1['dBS'].tolist()).reshape(-1,1)\n",
    "    corr, pval = stats.pearsonr(dbs, skew)\n",
    "\n",
    "    fig = plt.figure(figsize=(2,2), dpi=300)\n",
    "    ax = fig.add_subplot(111)\n",
    "    ax.set(ylim=(-4, 4))\n",
    "    ax.set(xlim=(-11, 11))\n",
    "    fig = plt.ylabel('LogSkew (Alt-Ref)', color='0.2', fontsize='xx-small')\n",
    "    fig = plt.xlabel('delta BS (Alt-Ref)', color='0.2',fontsize='xx-small')\n",
    "    fig =  plt.scatter(y=df1['LogSkew'], x=df1['dBS'], s=2, alpha=0.6,color=\"0.3\", zorder=1)\n",
    "    fig = plt.text(-10,-3,corr, color='0.2',fontsize='xx-small')\n",
    "\n",
    "    fig = plt.axhline(y=0, color='0.8', zorder=0)\n",
    "    fig = plt.axvline(x=0, color='0.8', zorder=0)\n",
    "\n",
    "    k='0.4'\n",
    "    ax.spines['right'].set_color(k)\n",
    "    ax.spines['left'].set_color(k)\n",
    "    ax.spines['top'].set_color(k)\n",
    "    ax.spines['bottom'].set_color(k)\n",
    "    ax.tick_params(axis='x', colors=k)\n",
    "    ax.tick_params(axis='y', colors=k)\n",
    "\n",
    "    fig =  plt.savefig('Fig5C_score_skew.%s.pdf' % segment)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
