{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aad465c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Figure 2D\n",
    "## Create piecewise linear plots to determine a motif score inflection point for each celltype enhancer combination\n",
    "\n",
    "library(ggplot2)\n",
    "library(reshape2)\n",
    "\n",
    "model_list <- list()\n",
    "for(celltype in unique(emvar_scores_motif_df$celltype)){\n",
    "  message(celltype)\n",
    "  model_list[[celltype]] <- list()\n",
    "  for(enh in enh_all){\n",
    "    message(enh)\n",
    "    model_list[[celltype]][[enh]]\n",
    "    cell_enh_score <- emvar_scores_motif_df$score[which(emvar_scores_motif_df$celltype==celltype & emvar_scores_motif_df$enhancer==enh)]\n",
    "    cell_enh_skew <- emvar_scores_motif_df$LogSkew[which(emvar_scores_motif_df$celltype==celltype & emvar_scores_motif_df$enhancer==enh)]\n",
    "    lin_mod <- lm(cell_enh_skew ~ cell_enh_score)\n",
    "    seg_mod <- segmented(lin_mod, seg.Z=~cell_enh_score, psi = c(20))\n",
    "    model_list[[celltype]][[enh]]$linear <- lin_mod\n",
    "    model_list[[celltype]][[enh]]$segmented <- seg_mod\n",
    "  }\n",
    "}\n",
    "\n",
    "reg_lines_plot_both <- function(celltype, enh, df){\n",
    "  group_temp <- df[which(df[,\"celltype\"]==celltype & df[,\"enhancer\"]==enh),c(\"celltype\",\"enhancer\",\"silencer\",\"LogSkew\",\"score\",\"skew_group\")]\n",
    "  \n",
    "  piecewise_fitted <- fitted(model_list[[celltype]][[enh]]$segmented)\n",
    "  linear_fitted <- fitted(model_list[[celltype]][[enh]]$linear)\n",
    "  \n",
    "  message(\"Fitted Models Complete\")\n",
    "  \n",
    "  p.model.temp <- data.frame(score=group_temp$score, LogSkew = piecewise_fitted)\n",
    "  l.model.temp <- data.frame(score=group_temp$score, LogSkew = linear_fitted)\n",
    "  \n",
    "  message(colnames(p.model.temp))\n",
    "  \n",
    "  message(\"DFs set up\")\n",
    "  p <- ggplot(group_temp, aes(x=score, y=LogSkew)) + geom_point(aes(col=skew_group), alpha=0.5, size=0.5) + xlim(0,35) + \n",
    "    geom_line(l.model.temp, mapping=aes(x=score, y=LogSkew), col=\"purple\", lwd=0.6) + \n",
    "    geom_line(p.model.temp, mapping=aes(x=score, y=LogSkew), col=\"salmon\", lwd=0.6) +\n",
    "    theme_light(base_size = 6) + theme(legend.position = \"none\", axis.title.y = element_text(size=4),\n",
    "                                       axis.title.x = element_text(size = 4), axis.text = element_text(size=3))\n",
    "  return(p)\n",
    "}\n",
    "\n",
    "for(celltype in names(model_list)){\n",
    "  message(celltype)\n",
    "  for(enh in names(model_list[[celltype]])){\n",
    "    message(enh)\n",
    "    p <- reg_lines_plot_both(celltype, enh, emvar_scores_motif_df)\n",
    "    \n",
    "    pdf(paste0(\"plots/piecewise_v_linear/\", celltype, \"_\", enh, \"_segmented.pdf\"), width = 45/25.4, height = 45/25.4)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0c4d852",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Figure 2E\n",
    "## Get lists of oligos which are above and below the motif inflection score\n",
    "\n",
    "oligo_split_list <- list()\n",
    "for(celltype in names(split_lower)){\n",
    "  oligo_split_list[[celltype]] <- list()\n",
    "  for(enh in names(split_lower[[celltype]])){\n",
    "    oligo_split_list[[celltype]][[enh]] <- list()\n",
    "    split_val <- 20.86\n",
    "    if(split_val < 25 & split_val > 15){\n",
    "      oligo_split_list[[celltype]][[enh]]$below <- emvar_scores_motif_df$silencer[which(emvar_scores_motif_df$celltype==celltype &\n",
    "                                                                                          emvar_scores_motif_df$enhancer==enh &\n",
    "                                                                                          emvar_scores_motif_df$score < split_val)]\n",
    "      oligo_split_list[[celltype]][[enh]]$above <- emvar_scores_motif_df$silencer[which(emvar_scores_motif_df$celltype==celltype &\n",
    "                                                                                          emvar_scores_motif_df$enhancer==enh &\n",
    "                                                                                          emvar_scores_motif_df$score >= split_val)]\n",
    "    }\n",
    "    if(split_val > 25 | split_val < 15){\n",
    "      oligo_split_list[[celltype]][[enh]]$below <- emvar_scores_motif_df$silencer[which(emvar_scores_motif_df$celltype==celltype &\n",
    "                                                                                          emvar_scores_motif_df$enhancer==enh )]\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "## Create box plots over swarm plots for each celltype enhancer combination of the log2FoldChange and motif positive/negative - score above and below categories\n",
    "## Calculate significance for each celltype enhancer combination and category in comparison to the negative controls\n",
    "\n",
    "scr_skew_stats <- list()\n",
    "for(enh in enh_all){\n",
    "  message(enh)\n",
    "  scale_max <- 9\n",
    "  if(enh==\"En02\"){\n",
    "    scale_min <- -6\n",
    "  }\n",
    "  if(enh==\"En09\"){\n",
    "    scale_min <- -2\n",
    "  }\n",
    "  if(enh==\"En11\"){\n",
    "    scale_min <- -5\n",
    "  }\n",
    "  if(enh==\"En19\"){\n",
    "    scale_min <- -2\n",
    "    scale_max <- 8\n",
    "  }\n",
    "  if(enh==\"En21\"){\n",
    "    scale_min <- 0\n",
    "    scale_max <- 10\n",
    "  }\n",
    "  to_plot <- as.data.frame(all_38_enh_res[[enh]])\n",
    "  to_plot <- to_plot[which(to_plot$proj %in% c(\"NegCtrl\",\"motif_ChIP_pos\",\"motif_ChIP_neg\")),]\n",
    "  # plot_split <- colsplit(rownames(to_plot),\"\\\\^\", c(\"enhancer\",\"silencer\"))\n",
    "  # sil_split <- colsplit(plot_split$silencer,\"\\\\.\", c(\"silencer\",\"rep\"))\n",
    "  # to_plot$sil <- sil_split$silencer\n",
    "  to_plot$group <- \"NoScore\"\n",
    "  to_plot$proj <- factor(to_plot$proj, levels = c(\"NegCtrl\",\"motif_ChIP_pos\",\"motif_ChIP_neg\"))\n",
    "  # to_plot$group <- factor(to_plot$group, levels = c(\"above\", \"below\", \"NoScore\"))\n",
    "  for(celltype in unique(to_plot$celltype)){\n",
    "    message(celltype)\n",
    "    for(type in names(oligo_split_list[[celltype]][[enh]])){\n",
    "      message(type)\n",
    "      to_plot$group[which(to_plot$sil %in% oligo_split_list[[celltype]][[enh]][[type]] & to_plot$celltype==celltype)] <- type\n",
    "    }\n",
    "  }\n",
    "  to_plot$group <- factor(to_plot$group, levels = c(\"above\", \"below\", \"NoScore\"))\n",
    "  to_plot$group[which(to_plot$proj!=\"NegCtrl\" & to_plot$group==\"NoScore\")] <- \"below\"\n",
    "  to_plot$inter <- interaction(to_plot$proj, to_plot$group)\n",
    "  to_plot$inter <- factor(to_plot$inter, levels = c(\"NegCtrl.NoScore\", \n",
    "                                                    \"motif_ChIP_pos.below\",\"motif_ChIP_pos.above\",\n",
    "                                                    \"motif_ChIP_neg.below\",\"motif_ChIP_neg.above\"))\n",
    "  scr_skew_stats[[enh]] <- list()\n",
    "  for(celltype in unique(to_plot$celltype)){\n",
    "    # message(celltype)\n",
    "    to_stat <- to_plot\n",
    "    to_stat <- remove.factors(to_stat)\n",
    "    stat_temp <- wilcox_test(to_stat[which(to_stat$celltype==celltype),], log2FoldChange~inter, ref.group = \"NegCtrl.NoScore\", p.adjust.method = \"BH\")\n",
    "    scr_skew_stats[[enh]][[celltype]] <- as.data.frame(stat_temp)\n",
    "  }\n",
    "  \n",
    "  med_plot <- to_plot[which(to_plot$proj==\"NegCtrl\"),] %>%\n",
    "    group_by(celltype) %>%\n",
    "    summarise(med_line = median(log2FoldChange, na.rm = T))\n",
    "  \n",
    "  pdf(paste0(\"plots/log2foldchange_box_plots/\",enh,\"_piecewise_box_swarm_alpha.pdf\"), width=174/25.4, height = (174/3.25)/25.4)\n",
    "  print(ggplot(to_plot, aes(x=inter, y=log2FoldChange, fill=inter, group=inter)) + geom_hline(data=med_plot, aes(yintercept=med_line)) + geom_quasirandom(aes(col=inter), outlier.shape=NA, size=0.5, method = \"smiley\", alpha=0.2) + geom_boxplot(outlier.shape = NA) +\n",
    "          ylim(scale_min,scale_max) + facet_grid(~celltype) + theme_light(base_size = 6) + xlab(\"\") + ylab(\"log2FoldChange\") +\n",
    "          theme(axis.text.x = element_blank(), legend.position = \"bottom\", strip.background = element_blank(), strip.text.x = element_blank(),legend.margin=margin(t = -.5, unit='cm'), panel.grid = element_line(size = 0.24), panel.border = element_rect(size = 0.24), legend.text = element_text(size=4)) +\n",
    "          guides(fill=guide_legend(nrow = 1), shape=guide_legend(override.aes = list(size = 0.5)), color=guide_legend(override.aes = list(size = 0.5))) +\n",
    "          scale_fill_manual(values = c(\"grey\",\"#66C2A5\",\"#FC8D62\",\"#8DA0CB\",\"#E78AC3\",\"#A6D854\",\"#FFD92F\")) + scale_color_manual(values = c(\"grey\",\"#66C2A5\",\"#FC8D62\",\"#8DA0CB\",\"#E78AC3\",\"#A6D854\",\"#FFD92F\")))\n",
    "  dev.off()\n",
    "}\n",
    "\n",
    "\n",
    "model_comp_sum$improvement <- model_comp_sum$seg_r.squared/model_comp_sum$lin_r.squared\n",
    "model_comp_sum$adj.imp <- model_comp_sum$seg_adj.r.squared/model_comp_sum$lin_adj.r.squared\n",
    "\n",
    "breakpoints_sd <- sd(model_comp_sum$seg_breakpoints)\n",
    "breakpoints_mean <- mean(model_comp_sum$seg_breakpoints)\n",
    "\n",
    "model_comp_sum[which(abs(model_comp_sum$seg_breakpoints - breakpoints_mean) < breakpoints_sd*2),]\n",
    "\n",
    "\n",
    "scr_skew_stats_df <- data.frame()\n",
    "for(enh in enh_all){\n",
    "  message(enh)\n",
    "  for(celltype in names(scr_skew_stats[[enh]])){\n",
    "    message(celltype)\n",
    "    temp <- data.frame(scr_skew_stats[[enh]][[celltype]])\n",
    "    temp$celltype <- celltype\n",
    "    temp$enhancer <- enh\n",
    "    scr_skew_stats_df <- rbind(scr_skew_stats_df,temp)\n",
    "  }\n",
    "}\n",
    "\n",
    "write.table(scr_skew_stats_df,\"results/skew_significance_cell_enh.txt\",quote = F, sep = \"\\t\",row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f1b7935",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Figure 4B\n",
    "## Link silencers with all TFs bound to that silencer\n",
    "\n",
    "tf_plots <- data.frame()\n",
    "for(celltype in names(standard_res)){\n",
    "  temp <- standard_res[[celltype]][standard_res[[celltype]]$project==\"motif\",c(\"ID\",\"log2FoldChange\")]\n",
    "  temp$celltype <- celltype\n",
    "  tf_plots <- rbind(tf_plots,temp)\n",
    "}\n",
    "\n",
    "tf_id_split <- colsplit(tf_plots$ID,\"\\\\^\",c(\"enhancer\",\"silencer\"))\n",
    "tf_plots$enhancer <- tf_id_split$enhancer\n",
    "tf_plots$silencer <- tf_id_split$silencer\n",
    "\n",
    "tf_plots$tf_stat <- 0\n",
    "for(celltype in unique(tf_plots$celltype)){\n",
    "  message(celltype)\n",
    "  for(sil in unique(tf_plots$silencer)){\n",
    "    if(any(chip_tf_celltypes[[celltype]][which(chip_tf_celltypes[[celltype]]$name==sil),5:ncol(chip_tf_celltypes[[celltype]])]==1)){\n",
    "      tf_plots$tf_stat[which(tf_plots$silencer==sil & tf_plots$celltype==celltype)] <- 1\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "tf_plots$tf_stat <- as.factor(tf_plots$tf_stat)\n",
    "tf_plots$celltype <- as.factor(tf_plots$celltype)\n",
    "\n",
    "accession_tf_name_map <- read.delim(\"ENCODE_ChIP/complete_histone_tf_list.txt\", stringsAsFactors = F)\n",
    "colnames(accession_tf_name_map) <- c(\"Accession\",\"Target\",\"Celltype\")\n",
    "\n",
    "accession_celltype_list <- list()\n",
    "accession_tf_name_map$Celltype[which(accession_tf_name_map$Celltype==\"SK-N-SH\")] <- \"SKNSH\"\n",
    "accession_tf_name_map$Celltype[which(accession_tf_name_map$Celltype==\"HEPG2\")] <- \"HepG2\"\n",
    "for(celltype in unique(accession_tf_name_map$Celltype)){\n",
    "  message(celltype)\n",
    "  accession_celltype_list[[celltype]] <- list()\n",
    "  for(tf in unique(accession_tf_name_map$Target[which(accession_tf_name_map$Celltype==celltype)])){\n",
    "    # if(tf==\"REST\") next\n",
    "    accession_celltype_list[[celltype]][[tf]] <- accession_tf_name_map$Accession[which(accession_tf_name_map$Target==tf & accession_tf_name_map$Celltype==celltype)]\n",
    "  }\n",
    "}\n",
    "\n",
    "## Count how many transcirption factors bind with each oligo\n",
    "\n",
    "bound_factor <- tf_plots[,1:5]\n",
    "bound_factor$REST <- \"unbound\"\n",
    "for(celltype in unique(bound_factor$celltype)){\n",
    "  message(celltype)\n",
    "  for(sil in unique(bound_factor$silencer[which(bound_factor$celltype==celltype)])){\n",
    "    if(isTRUE(chip_tf_celltypes[[celltype]][which(chip_tf_celltypes[[celltype]]$name==sil),accession_celltype_list[[celltype]]$REST[1]]==1) | isTRUE(chip_tf_celltypes[[celltype]][which(chip_tf_celltypes[[celltype]]$name==sil),accession_celltype_list[[celltype]]$REST[2]]==1)){\n",
    "      bound_factor$REST[which(bound_factor$silencer==sil & bound_factor$celltype==celltype)] <- \"bound\"\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "## Add allelic skew information\n",
    "\n",
    "scramble_skew <- list()\n",
    "for(celltype in names(emVAR_all)){\n",
    "  message(celltype)\n",
    "  intersect_list <- bound_factor$ID[bound_factor$ID[bound_factor$celltype==celltype] %in% emVAR_all[[celltype]]$ID]\n",
    "  scramble_skew[[celltype]] <- emVAR_all[[celltype]][emVAR_all[[celltype]]$ID %in% intersect_list,c(\"ID\",\"B_log2FC\",\"LogSkew\")]\n",
    "}\n",
    "\n",
    "bound_factor$scr_l2fc <- NA\n",
    "bound_factor$LogSkew <- NA\n",
    "for(celltype in names(scramble_skew)){\n",
    "  message(celltype)\n",
    "  # bound_factor[which(bound_factor$celltype==celltype),] <- merge(bound_factor[which(bound_factor$celltype==celltype),], scramble_skew[[celltype]], by=\"ID\", all=T)\n",
    "  for(id in scramble_skew[[celltype]]$ID){\n",
    "    bound_factor$scr_l2fc[which(bound_factor$celltype==celltype & bound_factor$ID==id)] <- scramble_skew[[celltype]]$B_log2FC[which(scramble_skew[[celltype]]$ID==id)]\n",
    "    bound_factor$LogSkew[which(bound_factor$celltype==celltype & bound_factor$ID==id)] <- scramble_skew[[celltype]]$LogSkew[which(scramble_skew[[celltype]]$ID==id)]\n",
    "  }\n",
    "}\n",
    "\n",
    "bound_factor$skew_na <- is.na(bound_factor$LogSkew)\n",
    "\n",
    "bound_factor$num_bound <- NA\n",
    "bound_tf_names <- list()\n",
    "for(celltype in names(chip_tf_celltypes)){\n",
    "  message(celltype)\n",
    "  bound_tf_names[[celltype]] <- list()\n",
    "  for(sil in unique(bound_factor$silencer[which(bound_factor$celltype==celltype)])){\n",
    "    acc_bound <- colnames(chip_tf_celltypes[[celltype]])[chip_tf_celltypes[[celltype]][which(chip_tf_celltypes[[celltype]]$name==sil),]==1]\n",
    "    tf_bound <- unique(accession_tf_name_map$Target[which(accession_tf_name_map$Accession %in% acc_bound)])\n",
    "    bound_tf_names[[celltype]][[sil]] <- tf_bound\n",
    "    bound_factor$num_bound[which(bound_factor$silencer==sil & bound_factor$celltype==celltype)] <- length(tf_bound)\n",
    "  }\n",
    "}\n",
    "\n",
    "# accession_tf_name_map <- read.delim(\"ENCODE_ChIP/complete_histone_tf_list.txt\", stringsAsFactors = F)\n",
    "# colnames(accession_tf_name_map) <- c(\"Accession\",\"Target\",\"Celltype\")\n",
    "# \n",
    "# accession_celltype_list <- list()\n",
    "# accession_tf_name_map$Celltype[which(accession_tf_name_map$Celltype==\"SK-N-SH\")] <- \"SK.N.SH\"\n",
    "# accession_tf_name_map$Celltype[which(accession_tf_name_map$Celltype==\"HEPG2\")] <- \"HepG2\"\n",
    "# for(celltype in unique(accession_tf_name_map$Celltype)){\n",
    "#   message(celltype)\n",
    "#   accession_celltype_list[[celltype]] <- list()\n",
    "#   for(tf in unique(accession_tf_name_map$Target[which(accession_tf_name_map$Celltype==celltype)])){\n",
    "#     # if(tf==\"REST\") next\n",
    "#     accession_celltype_list[[celltype]][[tf]] <- accession_tf_name_map$Accession[which(accession_tf_name_map$Target==tf & accession_tf_name_map$Celltype==celltype)]\n",
    "#   }\n",
    "# }\n",
    "\n",
    "## Reorganize list so that it's organized by celltype and transcription factor\n",
    "tf_sil_list <- list()\n",
    "for(celltype in names(accession_celltype_list)){\n",
    "  message(celltype)\n",
    "  tf_sil_list[[celltype]] <- list()\n",
    "  temp <- chip_tf_celltypes[[celltype]]\n",
    "  for(tf in names(accession_celltype_list[[celltype]])){\n",
    "    temp_sil_list <- c()\n",
    "    for(acc in accession_celltype_list[[celltype]][[tf]]){\n",
    "      temp_sil_list <- c(temp_sil_list, chip_tf_celltypes[[celltype]]$name[which(chip_tf_celltypes[[celltype]][,acc]==1)])\n",
    "    }\n",
    "    tf_sil_list[[celltype]][[tf]] <- temp_sil_list\n",
    "  }\n",
    "}\n",
    "\n",
    "## Group oligos for each celltype and enhancer in order to categorize where TFs align on the spectrum of REST/TF binding\n",
    "\n",
    "med_comp_list <- list()\n",
    "med_comp_list$A_B <- list()\n",
    "med_comp_list$A_D <- list()\n",
    "med_comp_list$C_B <- list()\n",
    "med_comp_list$C_D <- list()\n",
    "med_comp_list$other <- list()\n",
    "for(celltype in names(tf_sil_list)){\n",
    "  message(celltype)\n",
    "  med_comp_list$A_B[[celltype]] <- list()\n",
    "  med_comp_list$A_D[[celltype]] <- list()\n",
    "  med_comp_list$C_B[[celltype]] <- list()\n",
    "  med_comp_list$C_D[[celltype]] <- list()\n",
    "  med_comp_list$other[[celltype]] <- list()\n",
    "  cell_sub <- bound_factor[which(bound_factor$celltype==celltype),1:8]\n",
    "  for(enh in unique(cell_sub$enhancer)){\n",
    "    message(enh)\n",
    "    med_comp_list$A_B[[celltype]][[enh]] <- list()\n",
    "    med_comp_list$A_D[[celltype]][[enh]] <- list()\n",
    "    med_comp_list$C_B[[celltype]][[enh]] <- list()\n",
    "    med_comp_list$C_D[[celltype]][[enh]] <- list()\n",
    "    med_comp_list$other[[celltype]][[enh]] <- list()\n",
    "    med_comp_list$A_B[[celltype]][[enh]]$TFs <- c()\n",
    "    med_comp_list$A_D[[celltype]][[enh]]$TFs <- c()\n",
    "    med_comp_list$C_B[[celltype]][[enh]]$TFs <- c()\n",
    "    med_comp_list$C_D[[celltype]][[enh]]$TFs <- c()\n",
    "    med_comp_list$other[[celltype]][[enh]]$TFs <- c()\n",
    "    cell_enh_sub <- cell_sub[which(cell_sub$enhancer==enh),]\n",
    "    for(tf in names(tf_sil_list[[celltype]])){\n",
    "      temp_sil_list <- tf_sil_list[[celltype]][[tf]]\n",
    "      RB_TFB <- median(cell_enh_sub$log2FoldChange[which(cell_enh_sub$REST==\"bound\" & cell_enh_sub$silencer %in% temp_sil_list)], na.rm = T)\n",
    "      RU_TFB <- median(cell_enh_sub$log2FoldChange[which(cell_enh_sub$REST==\"unbound\" & cell_enh_sub$silencer %in% temp_sil_list)], na.rm = T)\n",
    "      RB_TFU <- median(cell_enh_sub$log2FoldChange[which(cell_enh_sub$REST==\"bound\" & cell_enh_sub$silencer %notin% temp_sil_list)], na.rm = T)\n",
    "      RU_TFU <- median(cell_enh_sub$log2FoldChange[which(cell_enh_sub$REST==\"unbound\" & cell_enh_sub$silencer %notin% temp_sil_list)], na.rm = T)\n",
    "      if(any(is.na(c(RB_TFB, RU_TFB)))){\n",
    "        med_comp_list$other[[celltype]][[enh]]$TFs <- c(med_comp_list$other[[celltype]][[enh]]$TFs,tf)\n",
    "        next\n",
    "      }\n",
    "      if(RB_TFB > RB_TFU & RU_TFB > RU_TFU){\n",
    "        med_comp_list$A_B[[celltype]][[enh]]$TFs <- c(med_comp_list$A_B[[celltype]][[enh]]$TFs,tf)\n",
    "        med_comp_list$A_B[[celltype]][[enh]]$med_skew <- median(cell_enh_sub$LogSkew[which(cell_enh_sub$silencer %in% temp_sil_list)], na.rm = T)\n",
    "      }\n",
    "      else if(RB_TFB > RB_TFU & RU_TFB < RU_TFU){\n",
    "        med_comp_list$A_D[[celltype]][[enh]]$TFs <- c(med_comp_list$A_D[[celltype]][[enh]]$TFs,tf)\n",
    "        med_comp_list$A_D[[celltype]][[enh]]$med_skew <- median(cell_enh_sub$LogSkew[which((cell_enh_sub$REST==\"bound\" & cell_enh_sub$silencer %in% temp_sil_list) | \n",
    "                                                                                             (cell_enh_sub$REST==\"unbound\" & cell_enh_sub$silencer %notin% temp_sil_list))], na.rm = T)\n",
    "      }\n",
    "      else if(RB_TFB < RB_TFU & RU_TFB > RU_TFU){\n",
    "        med_comp_list$C_B[[celltype]][[enh]]$TFs <- c(med_comp_list$C_B[[celltype]][[enh]]$TFs,tf)\n",
    "        med_comp_list$C_B[[celltype]][[enh]]$med_skew <- median(cell_enh_sub$LogSkew[which((cell_enh_sub$REST==\"bound\" & cell_enh_sub$silencer %notin% temp_sil_list) | \n",
    "                                                                                             (cell_enh_sub$REST==\"unbound\" & cell_enh_sub$silencer %in% temp_sil_list))], na.rm = T)\n",
    "      }\n",
    "      else if(RB_TFB < RB_TFU & RU_TFB < RU_TFU){\n",
    "        med_comp_list$C_D[[celltype]][[enh]]$TFs <- c(med_comp_list$C_D[[celltype]][[enh]]$TFs,tf)\n",
    "        med_comp_list$C_D[[celltype]][[enh]]$med_skew <- median(cell_enh_sub$LogSkew[which(cell_enh_sub$silencer %notin% temp_sil_list)], na.rm = T)\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "med_comp_plot_df <- data.frame()\n",
    "for(celltype in names(tf_sil_list)){\n",
    "  message(celltype)\n",
    "  cell_sub <- bound_factor[which(bound_factor$celltype==celltype),1:8]\n",
    "  for(enh in unique(cell_sub$enhancer)){\n",
    "    message(enh)\n",
    "    cell_enh_sub <- cell_sub[which(cell_sub$enhancer==enh),]\n",
    "    for(tf in names(tf_sil_list[[celltype]])){\n",
    "      temp_sil_list <- tf_sil_list[[celltype]][[tf]]\n",
    "      A <- cell_enh_sub$log2FoldChange[which(cell_enh_sub$REST==\"bound\" & cell_enh_sub$silencer %in% temp_sil_list)]\n",
    "      A_med <- median(A, na.rm = T)\n",
    "      B <- cell_enh_sub$log2FoldChange[which(cell_enh_sub$REST==\"unbound\" & cell_enh_sub$silencer %in% temp_sil_list)]\n",
    "      B_med <- median(B, na.rm = T)\n",
    "      C <- cell_enh_sub$log2FoldChange[which(cell_enh_sub$REST==\"bound\" & cell_enh_sub$silencer %notin% temp_sil_list)]\n",
    "      C_med <- median(C, na.rm = T)\n",
    "      D <- cell_enh_sub$log2FoldChange[which(cell_enh_sub$REST==\"unbound\" & cell_enh_sub$silencer %notin% temp_sil_list)]\n",
    "      D_med <- median(D, na.rm = T)\n",
    "      if(length(A[is.na(A)])==length(A)){\n",
    "        A <- c()\n",
    "      }\n",
    "      if(length(B[is.na(B)])==length(B)){\n",
    "        B <- c()\n",
    "      }\n",
    "      if(length(C[is.na(C)])==length(C)){\n",
    "        C <- c()\n",
    "      }\n",
    "      if(length(D[is.na(D)])==length(D)){\n",
    "        D <- c()\n",
    "      }\n",
    "      if(length(A) > 0 & length(C) > 0){\n",
    "        A_C_sig <- wilcox.test(A,C)$p.value\n",
    "      }\n",
    "      if(length(A) == 0 | length(C) == 0){\n",
    "        A_C_sig <- NA\n",
    "      }\n",
    "      if(length(B) > 0 & length(D) > 0){\n",
    "        B_D_sig <- wilcox.test(B,D)$p.value\n",
    "      }\n",
    "      if(length(B) == 0 | length(D) == 0){\n",
    "        B_D_sig <- NA\n",
    "      }\n",
    "      temp <- data.frame(celltype, enh, tf, A_med, B_med, C_med, D_med, A_C_sig, B_D_sig)\n",
    "      med_comp_plot_df <- rbind(med_comp_plot_df, temp)\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "med_comp_plot_df$delta_AC <- med_comp_plot_df$A_med - med_comp_plot_df$C_med\n",
    "med_comp_plot_df$delta_BD <- med_comp_plot_df$B_med - med_comp_plot_df$D_med\n",
    "\n",
    "med_comp_plot_df$A_C_adj <- p.adjust(med_comp_plot_df$A_C_sig, method = \"hochberg\")\n",
    "med_comp_plot_df$B_D_adj <- p.adjust(med_comp_plot_df$B_D_sig, method = \"hochberg\")\n",
    "\n",
    "med_comp_plot_df$significance <- NA\n",
    "med_comp_plot_df$significance[which(med_comp_plot_df$A_C_adj < 0.05 & med_comp_plot_df$B_D_adj < 0.05)] <- \"both\"\n",
    "med_comp_plot_df$significance[which(med_comp_plot_df$A_C_adj < 0.05 & med_comp_plot_df$B_D_adj > 0.05)] <- \"REST+ only\"\n",
    "med_comp_plot_df$significance[which(med_comp_plot_df$A_C_adj > 0.05 & med_comp_plot_df$B_D_adj < 0.05)] <- \"REST- only\"\n",
    "med_comp_plot_df$significance[which(med_comp_plot_df$A_C_adj > 0.05 & med_comp_plot_df$B_D_adj > 0.05)] <- \"neither\"\n",
    "med_comp_plot_df$significance <- factor(med_comp_plot_df$significance, levels = c(\"both\",\"REST+ only\",\"REST- only\",\"neither\"))\n",
    "\n",
    "## save plot as pdf or interactive plotly\n",
    "\n",
    "for(celltype in unique(med_comp_plot_df$celltype)){\n",
    "  message(celltype)\n",
    "  for(enh in enh_all){\n",
    "    message(enh)\n",
    "    cell_enh_sub <- med_comp_plot_df[which(med_comp_plot_df$celltype==celltype & med_comp_plot_df$enh==enh),]\n",
    "    max_ac <- max(cell_enh_sub$delta_AC, na.rm=T)\n",
    "    min_ac <- min(cell_enh_sub$delta_AC, na.rm=T)\n",
    "    max_bd <- max(cell_enh_sub$delta_BD, na.rm = T)\n",
    "    min_bd <- min(cell_enh_sub$delta_BD, na.rm = T)\n",
    "    x_max <- max(abs(max_ac), abs(min_ac))\n",
    "    y_max <- max(abs(max_bd), abs(min_bd))\n",
    "    # pdf(paste0(\"plots/tf_analysis/scatter_plots/\",celltype,\"_\",enh,\"_delta_bound_v_unbound_centered_bhadj.pdf\"), height=55/25.4, width = 55/25.4)\n",
    "    a <- ggplot(cell_enh_sub, aes(x=delta_AC, y=delta_BD, col=significance, key=tf)) + geom_point(aes(alpha=0.3), size=2) + geom_hline(yintercept = 0) + geom_vline(xintercept = 0) +\n",
    "      xlim(-x_max,x_max) + ylim(-y_max,y_max) + xlab(\"\") + ylab(NULL) +\n",
    "      theme_light(base_size = 6) + scale_color_manual(values = c(\"red\",\"dodgerblue\",\"green3\",\"black\"),name=\"Significance\") +\n",
    "      theme(legend.position = \"bottom\", legend.margin=margin(t = -0.6, unit='cm'),legend.justification = \"left\", legend.key.width = unit(0.6, \"mm\")) + \n",
    "      guides(fill=guide_legend(nrow = 1),shape=guide_legend(override.aes = list(size = 0.5)),color = guide_legend(override.aes = list(size = 0.5)),\n",
    "             alpha=F, size=F)\n",
    "    # print(a)\n",
    "    # dev.off()\n",
    "    \n",
    "    a_ly <- ggplotly(a)\n",
    "    htmlwidgets::saveWidget(as_widget(a_ly), paste0(\"plots/tf_analysis/scatter_plots/\",celltype,\"_\",enh,\"_delta_bound_v_unbound_centered_bhadj.html\"))\n",
    "    \n",
    "    # pdf(paste0(\"plots/tf_analysis/scatter_plots/\",celltype,\"_\",enh,\"_delta_bound_v_unbound.pdf\"), height=55/25.4, width = 55/25.4)\n",
    "    # print(ggplot(cell_enh_sub, aes(x=delta_AC, y=delta_BD, col=significance)) + geom_point(aes(alpha=0.3), size=2) + geom_hline(yintercept = 0) + geom_vline(xintercept = 0) +\n",
    "    #         xlim(-2.5,7.25) + ylim(-2.5,7.25) + xlab(\"\") + ylab(\"\") +\n",
    "    #         theme_light(base_size = 6) + scale_color_manual(values = c(\"red\",\"dodgerblue\",\"green3\",\"black\"),name=\"Significance\") +\n",
    "    #         theme(legend.position = \"bottom\", legend.margin=margin(t = -0.5, unit='cm'),legend.justification = \"left\", legend.key.width = unit(0.7, \"mm\")) + \n",
    "    #         guides(fill=guide_legend(nrow = 1),shape=guide_legend(override.aes = list(size = 0.5)),color = guide_legend(override.aes = list(size = 0.5)),\n",
    "    #                alpha=F, size=F)) \n",
    "    # dev.off()\n",
    "    \n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f54b56dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Figure 4C\n",
    "## Calculate the odds ratio of each of the categories of TFs\n",
    "\n",
    "all_data_single_df <- data.frame()\n",
    "for(celltype in names(standard_res)){\n",
    "  message(celltype)\n",
    "  \n",
    "  id_split <- colsplit(standard_res[[celltype]]$ID, \"\\\\^\",c(\"enhancer\",\"silencer\"))\n",
    "  \n",
    "  for(enh in unique(id_split$enhancer)){\n",
    "    message(enh)\n",
    "    \n",
    "    temp <- id_split$silencer[which(id_split$enhancer==enh)]\n",
    "    temp2 <- standard_res[[celltype]][which(id_split$enhancer==enh), c(\"project\",\"log2FoldChange\",\"lfcSE\")]\n",
    "    \n",
    "    temp <- cbind(temp, temp2)\n",
    "    \n",
    "    colnames(temp) <- c(\"silencer\",\"project\",paste0(celltype,\".\",enh,\".l2FC\"), paste0(celltype,\".\",enh,\".l2FC_SE\"))\n",
    "    \n",
    "    if(celltype == \"GM12878\" & enh == \"En02\"){\n",
    "      all_data_single_df <- temp\n",
    "    }\n",
    "    else{\n",
    "      all_data_single_df <- merge(all_data_single_df, temp, by=c(\"silencer\",\"project\"), all=T)\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "head(all_data_single_df)\n",
    "\n",
    "write.table(all_data_single_df,\"results/silencer_project_l2FC_l2FC.SE_all_cell_enh.txt\", quote=F, sep = \"\\t\", row.names = F)\n",
    "\n",
    "emvar_all_motif_scores <- data.frame()\n",
    "for(celltype in names(emVAR_all)){\n",
    "  message(celltype)\n",
    "  emvar_temp <- emVAR_all[[celltype]][is.na(emVAR_all[[celltype]]$ref_allele),]\n",
    "  emvar_temp$celltype <- celltype\n",
    "  emvar_all_motif_scores <- rbind(emvar_all_motif_scores, emvar_temp)\n",
    "}\n",
    "\n",
    "head(emvar_all_motif_scores)\n",
    "motif_all_IDs <- colsplit(emvar_all_motif_scores$ID, \"\\\\^\", c(\"enhancer\",\"silencer\"))\n",
    "emvar_all_motif_scores$enhancer <- motif_all_IDs$enhancer\n",
    "emvar_all_motif_scores$silencer <- motif_all_IDs$silencer\n",
    "\n",
    "emvar_all_motif_scores <- merge(emvar_all_motif_scores, ref_all_scores[which(ref_all_scores$start > 90 & ref_all_scores$stop < 112 & ref_all_scores$X.pattern.name==\"MA0138.2\"),], by.x=\"silencer\", by.y=\"sequence.name\", drop=F)\n",
    "\n",
    "head(emvar_all_motif_scores)\n",
    "emvar_all_motif_scores$skew_group <- NA\n",
    "emvar_all_motif_scores$skew_group[which(emvar_all_motif_scores$LogSkew < 0)] <- 1\n",
    "emvar_all_motif_scores$skew_group[which(emvar_all_motif_scores$LogSkew > 0)] <- 0\n",
    "\n",
    "\n",
    "names(emvar_all_motif_scores)\n",
    "\n",
    "emvar_scores_motif_df <- emvar_all_motif_scores[,c(\"skew_group\",\"A_Ctrl_Mean\",\"A_Exp_Mean\",\"A_log2FC\",\"A_log2FC_SE\",\"A_logP\",\"A_logPadj_BH\",\"A_logPadj_BF\",\n",
    "                                                   \"B_Ctrl_Mean\",\"B_Exp_Mean\",\"B_log2FC\",\"B_log2FC_SE\",\"B_logP\",\"B_logPadj_BH\",\"B_logPadj_BF\",\"LogSkew\",\n",
    "                                                   \"Skew_logP\",\"Skew_logFDR\",\"celltype\",\"enhancer\",\"silencer\",\"score\")]\n",
    "\n",
    "emvar_scores_motif_df <- emvar_scores_motif_df[complete.cases(emvar_scores_motif_df$skew_group),]\n",
    "emvar_scores_motif_df$skew_group <- as.numeric(emvar_scores_motif_df$skew_group)\n",
    "\n",
    "\n",
    "oligo_split_list <- list()\n",
    "for(celltype in names(standard_res)){\n",
    "  oligo_split_list[[celltype]] <- list()\n",
    "  for(enh in enh_all){\n",
    "    oligo_split_list[[celltype]][[enh]] <- list()\n",
    "    # split_val <- model_comp_sum$seg_breakpoints[which(model_comp_sum$celltype==celltype & model_comp_sum$enh==enh)]\n",
    "    split_val <- 20.86\n",
    "    if(split_val < 25 & split_val > 15){\n",
    "      oligo_split_list[[celltype]][[enh]]$below <- emvar_scores_motif_df$silencer[which(emvar_scores_motif_df$celltype==celltype &\n",
    "                                                                                          emvar_scores_motif_df$enhancer==enh &\n",
    "                                                                                          emvar_scores_motif_df$score < split_val)]\n",
    "      oligo_split_list[[celltype]][[enh]]$above <- emvar_scores_motif_df$silencer[which(emvar_scores_motif_df$celltype==celltype &\n",
    "                                                                                          emvar_scores_motif_df$enhancer==enh &\n",
    "                                                                                          emvar_scores_motif_df$score >= split_val)]\n",
    "    }\n",
    "    if(split_val > 25 | split_val < 15){\n",
    "      oligo_split_list[[celltype]][[enh]]$below <- emvar_scores_motif_df$silencer[which(emvar_scores_motif_df$celltype==celltype &\n",
    "                                                                                        emvar_scores_motif_df$enhancer==enh )]\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "### REDO with color relating to significance and positivity of odds ratio as opposed to the actual odds ratio value\n",
    "or_tf_df <- data.frame()\n",
    "for(celltype in unique(med_comp_plot_df$celltype)){\n",
    "  message(celltype)\n",
    "  for(enh in enh_all){\n",
    "    message(enh)\n",
    "    cell_enh_sub <- med_comp_plot_df[which(med_comp_plot_df$celltype==celltype & med_comp_plot_df$enh==enh),]\n",
    "    \n",
    "    cell_enh_sub2 <- bound_factor[which(bound_factor$celltype==celltype & bound_factor$enhancer==enh),1:8]\n",
    "    \n",
    "    cell_enh_sub$OR <- NA\n",
    "    cell_enh_sub$OR_sig <- NA\n",
    "    for(tf in names(tf_sil_list[[celltype]])){\n",
    "      temp_sil_list <- tf_sil_list[[celltype]][[tf]]\n",
    "      A <- cell_enh_sub2$log2FoldChange[which(cell_enh_sub2$REST==\"bound\" & cell_enh_sub2$silencer %in% temp_sil_list & cell_enh_sub2$silencer %in% oligo_split_list[[celltype]][[enh]]$above)]\n",
    "      \n",
    "      B <- cell_enh_sub2$log2FoldChange[which(cell_enh_sub2$REST==\"unbound\" & cell_enh_sub2$silencer %in% temp_sil_list & cell_enh_sub2$silencer %notin% oligo_split_list[[celltype]][[enh]]$above)]\n",
    "      \n",
    "      C <- cell_enh_sub2$log2FoldChange[which(cell_enh_sub2$REST==\"bound\" & cell_enh_sub2$silencer %notin% temp_sil_list & cell_enh_sub2$silencer %in% oligo_split_list[[celltype]][[enh]]$above)]\n",
    "      \n",
    "      D <- cell_enh_sub2$log2FoldChange[which(cell_enh_sub2$REST==\"unbound\" & cell_enh_sub2$silencer %notin% temp_sil_list & cell_enh_sub2$silencer %notin% oligo_split_list[[celltype]][[enh]]$above)]\n",
    "      \n",
    "      OR_above_mat <- matrix(c(length(A),length(B),length(C),length(D)), nrow = 2)\n",
    "      ft_above <- fisher.test(OR_above_mat)\n",
    "      OR_above_est <- as.numeric(ft_above$estimate)\n",
    "      OR_above_pval <- as.numeric(ft_above$p.value)\n",
    "      cell_enh_sub$OR[which(cell_enh_sub$tf==tf)] <- OR_above_est\n",
    "      cell_enh_sub$OR_sig[which(cell_enh_sub$tf==tf)] <- OR_above_pval\n",
    "    }\n",
    "    cell_enh_sub$log2OR <- log2(cell_enh_sub$OR)\n",
    "    or_tf_df <- rbind(or_tf_df, cell_enh_sub)\n",
    "    \n",
    "    max_ac <- max(cell_enh_sub$delta_AC, na.rm=T)\n",
    "    min_ac <- min(cell_enh_sub$delta_AC, na.rm=T)\n",
    "    max_bd <- max(cell_enh_sub$delta_BD, na.rm = T)\n",
    "    min_bd <- min(cell_enh_sub$delta_BD, na.rm = T)\n",
    "    x_max <- max(abs(max_ac), abs(min_ac))\n",
    "    y_max <- max(abs(max_bd), abs(min_bd))\n",
    "    pdf(paste0(\"plots/tf_analysis/scatter_plots/odds_ratio/\",celltype,\"_\",enh,\"_delta_bound_v_unbound_centered.pdf\"), height=55/25.4, width = 55/25.4)\n",
    "    print(ggplot(cell_enh_sub, aes(x=delta_AC, y=delta_BD, col=log2OR, shape=significance, key=tf)) + geom_point(aes(alpha=0.3), size=1) + geom_hline(yintercept = 0) + geom_vline(xintercept = 0) +\n",
    "      xlim(-x_max,x_max) + ylim(-y_max,y_max) + xlab(\"\") + ylab(\"\") +\n",
    "      theme_light(base_size = 6) + scale_shape_manual(values = c(\"both\"=15,\"REST+ only\"=18,\"REST- only\"=17, \"neither\"=16),name=\"Significance\") + scale_color_gradient(low=\"blue\",high=\"orange\") +\n",
    "      theme(legend.position = \"bottom\", legend.margin=margin(t = -0.5, unit='cm'),legend.justification = \"left\", legend.key.width = unit(0.7, \"mm\"), legend.box = \"vertical\") +\n",
    "      guides(shape=guide_legend(override.aes = list(size = 0.5)),color = guide_legend(override.aes = list(size = 0.5)),\n",
    "             alpha=F, size=F))\n",
    "    dev.off()\n",
    "    \n",
    "    pdf(paste0(\"plots/tf_analysis/scatter_plots/odds_ratio/\",celltype,\"_\",enh,\"_delta_bound_v_unbound.pdf\"), height=55/25.4, width = 55/25.4)\n",
    "    print(ggplot(cell_enh_sub, aes(x=delta_AC, y=delta_BD, col=log2OR, shape=significance, key=tf)) + geom_point(aes(alpha=0.3), size=1) + geom_hline(yintercept = 0) + geom_vline(xintercept = 0) +\n",
    "            xlim(-2.5,7.25) + ylim(-2.5,7.25) + xlab(\"\") + ylab(\"\") +\n",
    "            theme_light(base_size = 6) + scale_shape_manual(values = c(\"both\"=15,\"REST+ only\"=18,\"REST- only\"=17, \"neither\"=16),name=\"Significance\") + scale_color_gradient(low=\"blue\",high=\"orange\") +\n",
    "            theme(legend.position = \"bottom\", legend.margin=margin(t = -0.5, unit='cm'),legend.justification = \"left\", legend.key.width = unit(0.7, \"mm\"), legend.box = \"vertical\") +\n",
    "            guides(shape=guide_legend(override.aes = list(size = 0.5)),color = guide_legend(override.aes = list(size = 0.5)),\n",
    "                   alpha=F, size=F))\n",
    "    dev.off()\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "or_tf_df$quadrant <- 0\n",
    "or_tf_df$quadrant[which(or_tf_df$delta_AC > 0 & or_tf_df$delta_BD > 0)] <- 1\n",
    "or_tf_df$quadrant[which(or_tf_df$delta_AC < 0 & or_tf_df$delta_BD > 0)] <- 2\n",
    "or_tf_df$quadrant[which(or_tf_df$delta_AC < 0 & or_tf_df$delta_BD < 0)] <- 3\n",
    "or_tf_df$quadrant[which(or_tf_df$delta_AC > 0 & or_tf_df$delta_BD < 0)] <- 4\n",
    "\n",
    "or_tf_df$sig_or <- \">=0.05\"\n",
    "or_tf_df$sig_or[which(or_tf_df$OR_sig < 0.05)] <- \"<0.05\"\n",
    "\n",
    "## Create interactive plotly version, or pdf of Odds Ratio significance plot\n",
    "\n",
    "for(celltype in unique(or_tf_df$celltype)){\n",
    "  message(celltype)\n",
    "  for(enh in enh_all){\n",
    "    message(enh)\n",
    "    to_plot <- or_tf_df[which(or_tf_df$celltype==celltype & or_tf_df$enh==enh & or_tf_df$significance==\"both\"),]\n",
    "    to_plot$quadrant <- factor(to_plot$quadrant, levels=c(\"1\",\"2\",\"3\",\"4\"))\n",
    "    # pdf(paste0(\"plots/tf_analysis/jitter/\",celltype,\"_\",enh,\"_OR_sig_quadrant_both_only.pdf\"), height=55/25.4, width = 55/25.4)\n",
    "    \n",
    "    tmp <- ggplot(to_plot, aes(x=quadrant, y=log2OR, col=sig_or, key=tf)) + geom_jitter(alpha=0.4, size=0.6) + # geom_point(alpha=0.4, size=0.6) +\n",
    "            scale_color_manual(values=c(\">=0.05\"=\"black\",\"<0.05\"=\"red\"), name=\"OR Significance\") + theme_light(base_size = 6) + theme(legend.position = \"bottom\")\n",
    "    tmply <- ggplotly(tmp)\n",
    "    \n",
    "    htmlwidgets::saveWidget(tmply,paste0(\"plots/tf_analysis/jitter/plotly/\",celltype,\"_\",enh,\"_OR_sig_quadrant_both_only.html\"))\n",
    "    # dev.off()\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d738a8c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Figure 4D-E\n",
    "## Identify and compare how many TFs have their motif present in each oligo\n",
    "\n",
    "sky_df <- data.frame()\n",
    "for(celltype in unique(tf_plots$celltype)){\n",
    "  message(celltype)\n",
    "  cell_ID_split <- colsplit(standard_res[[celltype]]$ID, \"\\\\^\", c(\"enhancer\",\"silencer\"))\n",
    "  for(enh in unique(tf_plots$enhancer)){\n",
    "    message(enh)\n",
    "    for(tf in colnames(chip_tf_celltypes[[celltype]])[5:ncol(chip_tf_celltypes[[celltype]])]){\n",
    "      bound_l2fc <- tf_plots$log2FoldChange[tf_plots$silencer %in% chip_tf_celltypes[[celltype]]$name[which(chip_tf_celltypes[[celltype]][,colnames(chip_tf_celltypes[[celltype]])==tf]==1)] & tf_plots$celltype==celltype & tf_plots$enhancer==enh]\n",
    "      unbound_l2fc <- tf_plots$log2FoldChange[tf_plots$silencer %in% chip_tf_celltypes[[celltype]]$name[which(chip_tf_celltypes[[celltype]][,colnames(chip_tf_celltypes[[celltype]])==tf]==0)] & tf_plots$celltype==celltype & tf_plots$enhancer==enh]\n",
    "      bound_med <- median(bound_l2fc, na.rm = T)\n",
    "      unbound_med <- median(unbound_l2fc, na.rm = T)\n",
    "      if(sum(is.na(bound_l2fc))==length(bound_l2fc)){\n",
    "        bound_l2fc <- c()\n",
    "      }\n",
    "      if(sum(is.na(unbound_l2fc))==length(unbound_l2fc)){\n",
    "        unbound_l2fc <- c()\n",
    "      }\n",
    "      if(length(bound_l2fc) > 0 & length(unbound_l2fc) > 0){\n",
    "        p_value <- wilcox.test(bound_l2fc,unbound_l2fc)$p.value\n",
    "      }\n",
    "      if(length(bound_l2fc) == 0 | length(unbound_l2fc) ==0){\n",
    "        p_value <- NA\n",
    "      }\n",
    "      percent_bound <- sum(chip_tf_celltypes[[celltype]][,colnames(chip_tf_celltypes[[celltype]])==tf]==1)/nrow(chip_tf_celltypes[[celltype]])\n",
    "      temp <- data.frame(tf,celltype,enh,p_value,bound_med,unbound_med,percent_bound)\n",
    "      sky_df <- rbind(sky_df, temp)\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "sky_df <- sky_df[order(sky_df$p_value, decreasing = T),]\n",
    "rownames(sky_df) <- seq(1:nrow(sky_df))\n",
    "\n",
    "sky_df$tf <- factor(sky_df$tf, levels = sky_df$tf[order(sky_df$p_value, decreasing = T)])\n",
    "\n",
    "for(celltype in unique(sky_df$celltype)){\n",
    "  for(enh in unique(sky_df$enh)){\n",
    "    sub_df <- sky_df[which(sky_df$celltype==celltype & sky_df$enh==enh),]\n",
    "    sub_df$tf <- factor(sub_df$tf, levels = sub_df$tf[order(complete.cases(sub_df$p_value), decreasing = T)])\n",
    "    png(paste0(\"plots/tf_analysis/skyscraper_plots/neg_log10_p/tf_names_\",celltype,\"_\",enh,\"_neglog10p.png\"), width = 1620, height = 1620)\n",
    "    print(ggplot(sub_df, aes(x=Target, y=-log10(p_value))) + geom_point() + labs(title = paste0(\"skyscraper plot: \",celltype,\" - \",enh,\"\\n-log10(p)\")) + theme_light() + theme(axis.text.x = element_text(angle = 90)))\n",
    "    dev.off()\n",
    "  }\n",
    "}\n",
    "## Delta median of bound - unbound, color if MWU test has a significant p-value. Taking median of the l2FC for the delta median calculation\n",
    "\n",
    "sky_df$significance <- \"> 0.05\"\n",
    "sky_df$significance[which(sky_df$p_value < 0.05)] <- \"< 0.05\" \n",
    "\n",
    "sky_df$significance <- factor(sky_df$significance, levels = c(\"< 0.05\",\"> 0.05\"))\n",
    "\n",
    "sky_df$delta_med <- sky_df$bound_med - sky_df$unbound_med\n",
    "\n",
    "accession_tf_name_map <- read.delim(\"ENCODE_ChIP/complete_histone_tf_list.txt\", stringsAsFactors = F)\n",
    "colnames(accession_tf_name_map) <- c(\"Accession\",\"Target\",\"Celltype\")\n",
    "\n",
    "sky_df <- merge(sky_df,accession_tf_name_map[,1:2], by.x=\"tf\",by.y=\"Accession\", all=T)\n",
    "\n",
    "bind_num_tf_plots <- tf_plots[,1:5]\n",
    "\n",
    "bound_tf_names <- list()\n",
    "bind_num_tf_plots$TF_num <- 0\n",
    "for(celltype in names(chip_tf_celltypes)){\n",
    "  bound_tf_names[[celltype]] <- list()\n",
    "  message(celltype)\n",
    "  for(sil in unique(bind_num_tf_plots$silencer[which(bind_num_tf_plots$celltype==celltype)])){\n",
    "    acc_list <- colnames(chip_tf_celltypes[[celltype]])[chip_tf_celltypes[[celltype]][which(chip_tf_celltypes[[celltype]]$name==sil),]==1]\n",
    "    tf_list <- accession_tf_name_map$Target[which(accession_tf_name_map$Accession %in% acc_list)]\n",
    "    bound_tf_names[[celltype]][[sil]] <- unique(tf_list)\n",
    "    total_bound_not_rest <- sum(bound_tf_names[[celltype]][[sil]]!=\"REST\")\n",
    "    bind_num_tf_plots$TF_num[which(bind_num_tf_plots$silencer==sil & bind_num_tf_plots$celltype==celltype)] <- total_bound_not_rest\n",
    "  }\n",
    "}\n",
    "\n",
    "bound_tf_names_filt <- list()\n",
    "bind_num_tf_plots$TF_num_filt <- 0\n",
    "for(celltype in names(chip_tf_celltypes)){\n",
    "  bound_tf_names_filt[[celltype]] <- list()\n",
    "  message(celltype)\n",
    "  for(sil in unique(bind_num_tf_plots$silencer[which(bind_num_tf_plots$celltype==celltype)])){\n",
    "    acc_list <- colnames(chip_tf_celltypes[[celltype]])[chip_tf_celltypes[[celltype]][which(chip_tf_celltypes[[celltype]]$name==sil),]==1]\n",
    "    acc_list <- acc_list[acc_list %in% sky_df$tf[sky_df$percent_bound>0.1]]\n",
    "    tf_list <- accession_tf_name_map$Target[which(accession_tf_name_map$Accession %in% acc_list)]\n",
    "    bound_tf_names_filt[[celltype]][[sil]] <- unique(tf_list)\n",
    "    total_bound_not_rest <- sum(bound_tf_names_filt[[celltype]][[sil]]!=\"REST\")\n",
    "    bind_num_tf_plots$TF_num_filt[which(bind_num_tf_plots$silencer==sil & bind_num_tf_plots$celltype==celltype)] <- total_bound_not_rest\n",
    "  }\n",
    "}\n",
    "\n",
    "freq_table_filt <- list()\n",
    "for(celltype in unique(bind_num_tf_plots$celltype)){\n",
    "  message(celltype)\n",
    "  freq_table_filt[[celltype]] <- table(bind_num_tf_plots$TF_num_filt[which(bind_num_tf_plots$celltype==celltype)])\n",
    "}\n",
    "\n",
    "\n",
    "## Making the plots\n",
    "head(bind_num_tf_plots)\n",
    "\n",
    "bins12 <- c()\n",
    "for(i in 1:12){\n",
    "  bins12 <- c(bins12, rgb(red = 30/255, green = 144/255, blue = 255/255, alpha = i/12))\n",
    "}\n",
    "\n",
    "for(celltype in unique(bind_num_tf_plots$celltype)){\n",
    "  message(celltype)\n",
    "  neg_ctrls_temp <- standard_res[[celltype]][which(standard_res[[celltype]]$project==\"NegCtrl\"),c(\"ID\",\"log2FoldChange\")]\n",
    "  neg_ctrls_temp_split <- colsplit(neg_ctrls_temp$ID, \"\\\\^\", c(\"enhancer\",\"silencer\"))\n",
    "  neg_ctrls_temp$celltype <- celltype\n",
    "  neg_ctrls_temp$enhancer <- neg_ctrls_temp_split$enhancer\n",
    "  neg_ctrls_temp$silencer <- neg_ctrls_temp_split$silencer\n",
    "  neg_ctrls_temp$TF_num <- \"NegCtrl\"\n",
    "  neg_ctrls_temp$TF_num_filt <- \"NegCtrl\"\n",
    "  for(enh in unique(bind_num_tf_plots$enhancer)){\n",
    "    message(enh)\n",
    "    neg_ctrls_enh <- neg_ctrls_temp[which(neg_ctrls_temp$enhancer==enh),]\n",
    "    plot_temp <- rbind(bind_num_tf_plots[which(bind_num_tf_plots$celltype==celltype & bind_num_tf_plots$enhancer==enh),], neg_ctrls_enh)\n",
    "    plot_temp$non_rest_tf <- plot_temp$TF_num\n",
    "    plot_temp$non_rest_tf[which(as.numeric(plot_temp$TF_num) >5 & as.numeric(plot_temp$TF_num) <11)] <- \"6-10\"\n",
    "    plot_temp$non_rest_tf[which(as.numeric(plot_temp$TF_num) >10 & as.numeric(plot_temp$TF_num) <21)] <- \"11-20\"\n",
    "    plot_temp$non_rest_tf[which(as.numeric(plot_temp$TF_num) >20 & as.numeric(plot_temp$TF_num) <31)] <- \"21-30\"\n",
    "    plot_temp$non_rest_tf[which(as.numeric(plot_temp$TF_num) >30 & as.numeric(plot_temp$TF_num) <41)] <- \"31-40\"\n",
    "    plot_temp$non_rest_tf[which(as.numeric(plot_temp$TF_num) >40 & as.numeric(plot_temp$TF_num) <51)] <- \"41-50\"\n",
    "    plot_temp$non_rest_tf[which(as.numeric(plot_temp$TF_num) >50)] <- \"> 50\"\n",
    "    plot_temp$non_rest_tf_filt <- plot_temp$TF_num_filt\n",
    "    plot_temp$non_rest_tf_filt[which(as.numeric(plot_temp$TF_num_filt) >5 & as.numeric(plot_temp$TF_num_filt) <11)] <- \"6-10\"\n",
    "    plot_temp$non_rest_tf_filt[which(as.numeric(plot_temp$TF_num_filt) >10 & as.numeric(plot_temp$TF_num_filt) <21)] <- \"11-20\"\n",
    "    plot_temp$non_rest_tf_filt[which(as.numeric(plot_temp$TF_num_filt) >20 & as.numeric(plot_temp$TF_num_filt) <31)] <- \"21-30\"\n",
    "    plot_temp$non_rest_tf_filt[which(as.numeric(plot_temp$TF_num_filt) >30 & as.numeric(plot_temp$TF_num_filt) <41)] <- \"31-40\"\n",
    "    plot_temp$non_rest_tf_filt[which(as.numeric(plot_temp$TF_num_filt) >40 & as.numeric(plot_temp$TF_num_filt) <51)] <- \"41-50\"\n",
    "    plot_temp$non_rest_tf_filt[which(as.numeric(plot_temp$TF_num_filt) >50)] <- \"> 50\"\n",
    "    plot_temp$non_rest_tf <- factor(plot_temp$non_rest_tf, levels = c(0,1,2,3,4,5,\"6-10\",\"11-20\",\"21-30\",\"31-40\",\"41-50\",\"> 50\"))\n",
    "    plot_temp$non_rest_tf_filt <- factor(plot_temp$non_rest_tf_filt, levels = c(0,1,2,3,4,5,\"6-10\",\"11-20\",\"21-30\",\"31-40\",\"41-50\",\"> 50\"))\n",
    "    plot_temp$project <- \"motif\"\n",
    "    plot_temp$project[plot_temp$TF_num==\"NegCtrl\"] <- \"NegCtrl\"\n",
    "    plot_temp$project <- factor(plot_temp$project, levels = c(\"NegCtrl\",\"motif\"))\n",
    "    # pdf(paste0(\"plots/tf_analysis/box_plots/number_tf_bound/no_filter/l2fc_neg_ctrls_v_motif_\",celltype,\"_\",enh,\"_blue.pdf\"), width = 3.55, height = 3.55)\n",
    "    # print(ggplot(plot_temp, aes(x=project, y=log2FoldChange, group=interaction(project,non_rest_tf))) + geom_boxplot(aes(fill=non_rest_tf),lwd=.24) + labs(title = paste0(\"log2FoldChange NegCtrl vs Motif - Number of Non-REST Bound Transcription Factors\\n\", celltype,\" - \",enh)) + theme_light(base_size = 6) + scale_fill_manual(values = bins12, name=\"# of\\nTFs\"))\n",
    "    # dev.off()\n",
    "    pdf(paste0(\"plots/tf_analysis/box_plots/number_tf_bound/filtered/l2fc_neg_ctrls_v_motif_\",celltype,\"_\",enh,\"_blue.pdf\"), width = 3.55, height = 3.55/.8)\n",
    "    print(ggplot(plot_temp, aes(x=project, y=log2FoldChange, group=interaction(project,non_rest_tf_filt))) +\n",
    "            geom_boxplot(aes(fill=non_rest_tf_filt), lwd=.24, outlier.shape = NA) +\n",
    "            labs(title = paste0(\"log2FC NegCtrl vs Motif - # of Non-REST Bound TFs\\n\", celltype,\" - \",enh,\"\\tFiltered by Percent Bound\")) +\n",
    "            theme_light(base_size = 6) + theme(legend.position = \"bottom\",legend.margin=margin(t = 0, unit='cm'),legend.justification = \"left\") +\n",
    "            scale_fill_manual(values = bins12, name=\"# of\\nTFs\", na.translate=T, na.value=\"grey\") + guides(fill=guide_legend(nrow = 2)))\n",
    "    dev.off()\n",
    "  }\n",
    "}\n",
    "\n",
    "new_tf_num_plot_df <- for_rodrigo[,1:9]\n",
    "new_tf_num_plot_df$TF_num <- NA\n",
    "new_tf_num_plot_df$TF_A_B <- NA\n",
    "new_tf_num_plot_df$TF_C_B <- NA\n",
    "for(celltype in unique(new_tf_num_plot_df$celltype)){\n",
    "  message(celltype)\n",
    "  for(sil in unique(new_tf_num_plot_df$silencer[which(new_tf_num_plot_df$celltype==celltype)])){\n",
    "    if(new_tf_num_plot_df$project[which(new_tf_num_plot_df$silencer==sil & new_tf_num_plot_df$celltype==celltype)][1]==\"NegCtrl\") next\n",
    "    total_bound_not_rest <- sum(bound_tf_names[[celltype]][[sil]]!=\"REST\")\n",
    "    new_tf_num_plot_df$TF_num[which(new_tf_num_plot_df$celltype==celltype & new_tf_num_plot_df$silencer==sil)] <-total_bound_not_rest\n",
    "    for(enh in unique(new_tf_num_plot_df$enhancer)){\n",
    "      total_bound_not_rest_A_B <- sum(bound_tf_names[[celltype]][[sil]]!=\"REST\" & bound_tf_names[[celltype]][[sil]] %in% med_comp_plot_df$tf[which(med_comp_plot_df$celltype==celltype & med_comp_plot_df$enh==enh & med_comp_plot_df$significance==\"both\" & med_comp_plot_df$quadrant==1)])\n",
    "      total_bound_not_rest_C_B <- sum(bound_tf_names[[celltype]][[sil]]!=\"REST\" & bound_tf_names[[celltype]][[sil]] %in% med_comp_plot_df$tf[which(med_comp_plot_df$celltype==celltype & med_comp_plot_df$enh==enh & med_comp_plot_df$significance==\"both\" & med_comp_plot_df$quadrant==2)])\n",
    "      new_tf_num_plot_df$TF_A_B[which(new_tf_num_plot_df$celltype==celltype & new_tf_num_plot_df$silencer==sil & new_tf_num_plot_df$enhancer==enh)] <-total_bound_not_rest_A_B\n",
    "      new_tf_num_plot_df$TF_C_B[which(new_tf_num_plot_df$celltype==celltype & new_tf_num_plot_df$silencer==sil & new_tf_num_plot_df$enhancer==enh)] <-total_bound_not_rest_C_B\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "new_tf_num_plot_df$tf_bins <- new_tf_num_plot_df$TF_num\n",
    "new_tf_num_plot_df$tf_bins[which(as.numeric(new_tf_num_plot_df$TF_num) > 5 & as.numeric(new_tf_num_plot_df$TF_num) < 11)] <- \"6-10\"\n",
    "new_tf_num_plot_df$tf_bins[which(as.numeric(new_tf_num_plot_df$TF_num) > 10 & as.numeric(new_tf_num_plot_df$TF_num) < 21)] <- \"11-20\"\n",
    "new_tf_num_plot_df$tf_bins[which(as.numeric(new_tf_num_plot_df$TF_num) > 20 & as.numeric(new_tf_num_plot_df$TF_num) < 31)] <- \"21-30\"\n",
    "new_tf_num_plot_df$tf_bins[which(as.numeric(new_tf_num_plot_df$TF_num) > 30 & as.numeric(new_tf_num_plot_df$TF_num) < 41)] <- \"31-40\"\n",
    "new_tf_num_plot_df$tf_bins[which(as.numeric(new_tf_num_plot_df$TF_num) > 40 & as.numeric(new_tf_num_plot_df$TF_num) < 51)] <- \"41-50\"\n",
    "new_tf_num_plot_df$tf_bins[which(new_tf_num_plot_df$TF_num > 50)] <- \">50\"\n",
    "\n",
    "new_tf_num_plot_df$tf_bins <- factor(new_tf_num_plot_df$tf_bins, levels = c(\"NegCtrl\",0,1,2,3,4,5,\"6-10\",\"11-20\",\"21-30\",\"31-40\",\"41-50\",\">50\"))\n",
    "new_tf_num_plot_df$tf_bins[is.na(new_tf_num_plot_df$tf_bins)] <- \"NegCtrl\"\n",
    "\n",
    "new_tf_num_plot_df$tf_bins_A_B <- new_tf_num_plot_df$TF_A_B\n",
    "new_tf_num_plot_df$tf_bins_A_B[which(as.numeric(new_tf_num_plot_df$TF_A_B) > 5 & as.numeric(new_tf_num_plot_df$TF_A_B) < 11)] <- \"6-10\"\n",
    "new_tf_num_plot_df$tf_bins_A_B[which(as.numeric(new_tf_num_plot_df$TF_A_B) > 10 & as.numeric(new_tf_num_plot_df$TF_A_B) < 21)] <- \"11-20\"\n",
    "new_tf_num_plot_df$tf_bins_A_B[which(as.numeric(new_tf_num_plot_df$TF_A_B) > 20 & as.numeric(new_tf_num_plot_df$TF_A_B) < 31)] <- \"21-30\"\n",
    "new_tf_num_plot_df$tf_bins_A_B[which(as.numeric(new_tf_num_plot_df$TF_A_B) > 30 & as.numeric(new_tf_num_plot_df$TF_A_B) < 41)] <- \"31-40\"\n",
    "new_tf_num_plot_df$tf_bins_A_B[which(as.numeric(new_tf_num_plot_df$TF_A_B) > 40 & as.numeric(new_tf_num_plot_df$TF_A_B) < 51)] <- \"41-50\"\n",
    "new_tf_num_plot_df$tf_bins_A_B[which(new_tf_num_plot_df$TF_A_B > 50)] <- \">50\"\n",
    "\n",
    "new_tf_num_plot_df$tf_bins_A_B <- factor(new_tf_num_plot_df$tf_bins_A_B, levels = c(\"NegCtrl\",0,1,2,3,4,5,\"6-10\",\"11-20\",\"21-30\",\"31-40\",\"41-50\",\">50\"))\n",
    "new_tf_num_plot_df$tf_bins_A_B[which(new_tf_num_plot_df$project==\"NegCtrl\")] <- \"NegCtrl\"\n",
    "\n",
    "new_tf_num_plot_df$tf_bins_C_B <- new_tf_num_plot_df$TF_C_B\n",
    "new_tf_num_plot_df$tf_bins_C_B[which(as.numeric(new_tf_num_plot_df$TF_C_B) > 5 & as.numeric(new_tf_num_plot_df$TF_C_B) < 11)] <- \"6-10\"\n",
    "new_tf_num_plot_df$tf_bins_C_B[which(as.numeric(new_tf_num_plot_df$TF_C_B) > 10 & as.numeric(new_tf_num_plot_df$TF_C_B) < 16)] <- \"11-15\"\n",
    "new_tf_num_plot_df$tf_bins_C_B[which(as.numeric(new_tf_num_plot_df$TF_C_B) > 15)] <- \">15\"\n",
    "\n",
    "new_tf_num_plot_df$tf_bins_C_B <- factor(new_tf_num_plot_df$tf_bins_C_B, levels = c(\"NegCtrl\",0,1,2,3,4,5,\"6-10\",\"11-15\",\">15\"))\n",
    "new_tf_num_plot_df$tf_bins_C_B[which(new_tf_num_plot_df$project==\"NegCtrl\")] <- \"NegCtrl\"\n",
    "\n",
    "\n",
    "\n",
    "bins_group <- data.frame()\n",
    "for(celltype in unique(new_tf_num_plot_df$celltype)){\n",
    "  message(celltype)\n",
    "  for(enh in unique(new_tf_num_plot_df$enhancer)){\n",
    "    message(enh)\n",
    "    cell_enh_sub <- new_tf_num_plot_df[which(new_tf_num_plot_df$enhancer==enh & new_tf_num_plot_df$celltype==celltype),]\n",
    "    neg_ctrls_enh <- cell_enh_sub[which(cell_enh_sub$project==\"NegCtrl\"),]\n",
    "    neg_ctrls_med <- median(neg_ctrls_enh$log2FoldChange, na.rm = T)\n",
    "    for(quad in names(new_num_tf_plot_info)){\n",
    "      message(quad)\n",
    "      plot_temp <- cell_enh_sub[which(cell_enh_sub$silencer %in% new_num_tf_plot_info[[quad]][[celltype]][[enh]]),]\n",
    "      if(nrow(plot_temp) == 0) next\n",
    "      if(quad==\"1\"){\n",
    "        group <- \"A_B\"\n",
    "        # skew_plot <- ggplot(plot_temp, aes(x=tf_bins_A_B, y=LogSkew, fill=tf_bins_A_B)) + geom_boxplot(outlier.shape = NA) + scale_fill_manual(values=bins12) + theme_light() +\n",
    "        #   theme(legend.position = \"none\") +xlab(\"Number of TFs\")\n",
    "        # skew_line <- ggplot(plot_temp, aes(x=tf_bins_A_B, y=LogSkew, fill=tf_bins_A_B)) + geom_hline(yintercept = 0, col=\"red\") + geom_boxplot(outlier.shape = NA) + scale_fill_manual(values=bins12) + theme_light() +\n",
    "        #   theme(legend.position = \"none\") +xlab(\"Number of TFs\")\n",
    "        # bound_stats <- data.frame(table(plot_temp[,c(\"tf_bins_A_B\",\"REST\")]))\n",
    "      }\n",
    "      if(quad==\"2\"){\n",
    "        group <- \"C_B\"\n",
    "        # skew_plot <- ggplot(plot_temp, aes(x=tf_bins_C_B, y=LogSkew, fill=tf_bins_C_B)) + geom_boxplot(outlier.shape = NA) + scale_fill_manual(values=bins12) + theme_light() +\n",
    "        #   theme(legend.position = \"none\") +xlab(\"Number of TFs\")\n",
    "        # skew_line <- ggplot(plot_temp, aes(x=tf_bins_C_B, y=LogSkew, fill=tf_bins_C_B)) + geom_hline(yintercept = 0, col=\"red\") + geom_boxplot(outlier.shape = NA) + scale_fill_manual(values=bins12) + theme_light() +\n",
    "        #   theme(legend.position = \"none\") +xlab(\"Number of TFs\")\n",
    "        # bound_stats <- data.frame(table(plot_temp[,c(\"tf_bins_C_B\",\"REST\")]))\n",
    "      }\n",
    "      \n",
    "      # bound_stats$percent <- NA\n",
    "      # for(i in 1:nrow(bound_stats)){\n",
    "      #   bin_sum <- sum(bound_stats$Freq[which(bound_stats[,1]==bound_stats[i,1])])\n",
    "      #   if(bin_sum==0) next\n",
    "      #   else{\n",
    "      #     bound_stats$percent[i] <- round(bound_stats$Freq[i]/bin_sum,digits = 4)\n",
    "      #   }\n",
    "      # }\n",
    "      # colnames(bound_stats) <- c(paste0(\"tf_bins_\",group),\"REST\",\"Freq\",\"percent\")\n",
    "      # if(quad==\"1\"){\n",
    "      #   bound_plot <- ggplot(bound_stats[complete.cases(bound_stats),], aes(x=tf_bins_A_B, y=Freq, fill=REST)) + geom_bar(stat=\"identity\") + theme_light(base_size = 6) +\n",
    "      #   geom_text(aes(label = percent), position = position_stack(vjust = 0.5), size=2) + scale_fill_manual(values = brewer_set2[1:3]) + theme(legend.position = \"bottom\",legend.box = \"vertical\",legend.margin=margin(t = 0, unit='cm'), \n",
    "      #                                                                                                                                  panel.grid = element_line(size = 0.24),legend.key.height = unit(2,'mm'), legend.key.width = unit(4,'mm'), legend.text.align = 0) +\n",
    "      #   xlab(\"Number of TFs\") + guides(fill=guide_legend(nrow = 1), title=\"REST binding\") + ylab(\"Number of Oligos\")\n",
    "      # }\n",
    "      # if(quad==\"2\"){\n",
    "      #   bound_plot <- ggplot(bound_stats[complete.cases(bound_stats),], aes(x=tf_bins_C_B, y=Freq, fill=REST)) + geom_bar(stat=\"identity\") + theme_light(base_size = 6) +\n",
    "      #     geom_text(aes(label = percent), position = position_stack(vjust = 0.5), size=2) + scale_fill_manual(values = brewer_set2[1:3]) + theme(legend.position = \"bottom\",legend.box = \"vertical\",legend.margin=margin(t = 0, unit='cm'), \n",
    "      #                                                                                                                                            panel.grid = element_line(size = 0.24),legend.key.height = unit(2,'mm'), legend.key.width = unit(4,'mm'), legend.text.align = 0) +\n",
    "      #     xlab(\"Number of TFs\") + guides(fill=guide_legend(nrow = 1), title=\"REST binding\") + ylab(\"Number of Oligos\")\n",
    "      # }\n",
    "      plot_tmp2 <- rbind(neg_ctrls_enh, plot_temp)\n",
    "      if(quad==\"1\"){\n",
    "        fc_plot <- ggplot(plot_tmp2, aes(x=tf_bins_A_B, y=log2FoldChange, fill=tf_bins_A_B)) + geom_hline(yintercept = neg_ctrls_med, col=\"red\") + geom_boxplot(outlier.shape = NA) + scale_fill_manual(values=c(\"grey\",bins12)) +\n",
    "          theme_light() + theme(legend.position = \"none\")  + xlab(\"Number of TFs\")\n",
    "      }\n",
    "      if(quad==\"2\"){\n",
    "        fc_plot <- ggplot(plot_tmp2, aes(x=tf_bins_C_B, y=log2FoldChange, fill=tf_bins_C_B)) + geom_hline(yintercept = neg_ctrls_med, col=\"red\") + geom_boxplot(outlier.shape = NA) + scale_fill_manual(values=c(\"grey\",bins12)) +\n",
    "          theme_light() + theme(legend.position = \"none\")  + xlab(\"Number of TFs\")\n",
    "      }\n",
    "      if(celltype==\"GM12878\" & enh==\"En02\" & quad==\"1\"){\n",
    "        plot_tmp2$group <- group\n",
    "        bins_group <- plot_tmp2\n",
    "      }\n",
    "      else{\n",
    "        if(nrow(plot_temp)>0){\n",
    "          plot_tmp2$group <- group\n",
    "          bins_group <- rbind(bins_group, plot_tmp2)\n",
    "        }\n",
    "      }\n",
    "      # bound_stats <- bound_stats[complete.cases(bound_stats),]\n",
    "      # if(quad==\"1\"){\n",
    "      #   bound_skew_plot <- ggplot(bound_stats[which(bound_stats$REST==1),], aes(x=tf_bins_A_B, y=percent, col=brewer_set2[2], group=1)) + geom_point() + scale_y_continuous(position = \"right\", limits = c(0,1), name=\"Fraction REST Bound\") +\n",
    "      #     geom_line() + theme_light(base_size = 6) + theme(legend.position = \"none\", axis.title.y.right = element_text(color = brewer_set2[2])) + xlab(\"Number of TFs\") \n",
    "      # }\n",
    "      # if(quad==\"2\"){\n",
    "      #   bound_skew_plot <- ggplot(bound_stats[which(bound_stats$REST==1),], aes(x=tf_bins_C_B, y=percent, col=brewer_set2[2], group=1)) + geom_point() + scale_y_continuous(position = \"right\", limits = c(0,1), name=\"Fraction REST Bound\") +\n",
    "      #     geom_line() + theme_light(base_size = 6) + theme(legend.position = \"none\", axis.title.y.right = element_text(color = brewer_set2[2])) + xlab(\"Number of TFs\") \n",
    "      # }\n",
    "      # message(\"plotting skew\")\n",
    "      # pdf(paste0(\"plots/tf_analysis/box_plots/skew_\",group,\"_\",enh,\"_\",celltype,\".pdf\"), width=170/25.4, height = 50/25.4)\n",
    "      # print(skew_plot)\n",
    "      # dev.off()\n",
    "      # message(\"plotting percent\")\n",
    "      # pdf(paste0(\"plots/tf_analysis/bar_plots/percent_bound_\",group,\"_\",enh,\"_\",celltype,\".pdf\"), width=170/25.4, height = 40/25.4)\n",
    "      # print(bound_plot)\n",
    "      # dev.off()\n",
    "      message(\"plotting l2fc\")\n",
    "      pdf(paste0(\"plots/tf_analysis/box_plots/l2fc_square_\",group,\"_\",enh,\"_\",celltype,\".pdf\"), width=40/25.4, height = 40/25.4)\n",
    "      print(fc_plot)\n",
    "      dev.off()\n",
    "      # message(\"plotting skew 2.0\")\n",
    "      # pdf(paste0(\"plots/tf_analysis/overlap_plots/skew_line_\",group,\"_\",enh,\"_\",celltype,\".pdf\"), width=40/25.4, height = 40/25.4)\n",
    "      # print(skew_line + theme_light(base_size = 6) + theme(axis.title.y = element_text(color = \"dodgerblue\"), legend.position = \"none\"))\n",
    "      # dev.off()\n",
    "      # message(\"plotting skew bound\")\n",
    "      # pdf(paste0(\"plots/tf_analysis/overlap_plots/bound_skew_\",group,\"_\",enh,\"_\",celltype,\".pdf\"), width = 40/25.4, height = 40/25.4)\n",
    "      # print(bound_skew_plot)\n",
    "      # dev.off()\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "pval_group_list <- list()\n",
    "for(celltype in unique(bins_group$celltype)){\n",
    "  message(celltype)\n",
    "  pval_group_list[[celltype]] <- list()\n",
    "  for(enh in unique(bins_group$enhancer)){\n",
    "    message(enh)\n",
    "    pval_group_list[[celltype]][[enh]] <- list()\n",
    "    for(group in unique(bins_group$group)){\n",
    "      message(group)\n",
    "      group_enh_cell_sub <- bins_group[which(bins_group$celltype==celltype & bins_group$enhancer==enh & bins_group$group==group),]\n",
    "      if(nrow(group_enh_cell_sub)==0) next\n",
    "      pval_group_list[[celltype]][[enh]][[group]] <- list()\n",
    "      group_enh_cell_sub <- remove.factors(group_enh_cell_sub)\n",
    "      if(group==\"A_B\"){\n",
    "        message(\"l2fc\")\n",
    "        pvals_l2fc <- pairwise_wilcox_test(group_enh_cell_sub, log2FoldChange ~ tf_bins_A_B, ref.group = \"NegCtrl\", p.adjust.method = \"BH\")\n",
    "        message(\"skew\")\n",
    "        group_enh_cell_sub <- group_enh_cell_sub[complete.cases(group_enh_cell_sub$LogSkew),]\n",
    "        pvals_skew <- pairwise_wilcox_test(group_enh_cell_sub, LogSkew ~ tf_bins_A_B, ref.group = \"1\", p.adjust.method = \"BH\")\n",
    "      }\n",
    "      if(group==\"C_B\"){\n",
    "        message(\"l2fc\")\n",
    "        groups <- as.data.frame(table(group_enh_cell_sub$tf_bins_C_B))\n",
    "        pvals_l2fc <- pairwise_wilcox_test(group_enh_cell_sub, log2FoldChange ~ tf_bins_C_B, ref.group = \"NegCtrl\", p.adjust.method = \"BH\")\n",
    "        message(\"skew\")\n",
    "        group_enh_cell_sub <- group_enh_cell_sub[complete.cases(group_enh_cell_sub$LogSkew),]\n",
    "        pvals_skew <- pairwise_wilcox_test(group_enh_cell_sub, LogSkew ~ tf_bins_C_B, ref.group = \"1\", p.adjust.method = \"BH\")\n",
    "      }\n",
    "      pval_group_list[[celltype]][[enh]][[group]]$l2fc <- pvals_l2fc\n",
    "      pval_group_list[[celltype]][[enh]][[group]]$skew <- pvals_skew\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02ee6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Figure F-I\n",
    "## Generate box plots to visualize the differences between Negative Controls and combinations of REST/MIER1/EHMT2\n",
    "## and REST/RCOR1/HDAC2 binding in oligos\n",
    "\n",
    "neg_ctrl_all <- data.frame()\n",
    "for(celltype in names(standard_res)){\n",
    "  message(celltype)\n",
    "  neg_ctrl_temp <- standard_res[[celltype]][which(standard_res[[celltype]]$project==\"NegCtrl\"),c(\"ID\",\"log2FoldChange\",\"project\")]\n",
    "  neg_ctrl_temp$celltype <- celltype\n",
    "  neg_ctrl_temp$scr_l2fc <- NA\n",
    "  neg_ctrl_temp$LogSkew <- NA\n",
    "  neg_ctrl_temp$REST <- \"NegCtrl\"\n",
    "  neg_ctrl_all <- rbind(neg_ctrl_all, neg_ctrl_temp)\n",
    "}\n",
    "\n",
    "neg_id_split <- colsplit(neg_ctrl_all$ID,\"\\\\^\",c(\"enhancer\",\"silencer\"))\n",
    "neg_ctrl_all$enhancer <- neg_id_split$enhancer\n",
    "neg_ctrl_all$silencer <- neg_id_split$silencer\n",
    "neg_ctrl_all$REST <- 0\n",
    "\n",
    "bound_factor <- merge(bound_factor, RESTscreen_attr_comb[,c(\"ID\",\"project\")], by=\"ID\", all.x=T)\n",
    "fig3 <- bound_factor[,c(\"ID\",\"enhancer\",\"silencer\",\"celltype\",\"log2FoldChange\",\"scr_l2fc\",\"LogSkew\",\"project\",\"REST\")]\n",
    "fig3 <- rbind(neg_ctrl_all, fig3)\n",
    "\n",
    "## Set up motif positivity for use in plots\n",
    "\n",
    "fig3$RCOR1 <- \"-\"\n",
    "fig3$REST <- \"-\"\n",
    "fig3$HDAC2 <- \"-\"\n",
    "fig3$MIER <- \"-\"\n",
    "fig3$EHMT2 <- \"-\"\n",
    "for(celltype in unique(fig3$celltype)){\n",
    "  message(celltype)\n",
    "  fig3$REST[which(fig3$celltype==celltype & fig3$silencer %in% tf_sil_list[[celltype]]$REST)] <- \"+\"\n",
    "  fig3$RCOR1[which(fig3$celltype==celltype & fig3$silencer %in% tf_sil_list[[celltype]]$RCOR1)] <- \"+\"\n",
    "  fig3$HDAC2[which(fig3$celltype==celltype & fig3$silencer %in% tf_sil_list[[celltype]]$HDAC2)] <- \"+\"\n",
    "  fig3$MIER[which(fig3$celltype==celltype & fig3$silencer %in% tf_sil_list[[celltype]]$MIER1)] <- \"+\"\n",
    "  fig3$EHMT2[which(fig3$celltype==celltype & fig3$silencer %in% tf_sil_list[[celltype]]$EHMT2)] <- \"+\"\n",
    "}\n",
    "\n",
    "fig3$all <- interaction(fig3$REST,fig3$MIER,fig3$EHMT2,fig3$RCOR1,fig3$HDAC2, sep=\"/\")\n",
    "\n",
    "fig3$REST_RCOR_HDAC <- interaction(fig3$REST,fig3$RCOR1,fig3$HDAC2, sep = \"/\")\n",
    "fig3$REST_MIER_EHMT2 <- interaction(fig3$REST,fig3$MIER,fig3$EHMT2, sep = \"/\")\n",
    "\n",
    "fig3 <- remove.factors(fig3)\n",
    "fig3$all[which(fig3$project==\"NegCtrl\")] <- \"NegCtrl\"\n",
    "fig3$REST_MIER_EHMT2[which(fig3$project==\"NegCtrl\")] <- \"NegCtrl\"\n",
    "fig3$REST_RCOR_HDAC[which(fig3$project==\"NegCtrl\")] <- \"NegCtrl\"\n",
    "\n",
    "## Run plots and get statistics through a Wilcox Test comparing the significance between each motif positivity group and the negative controls or REST/TF1/TF2 negative for skew\n",
    "\n",
    "# inter_stats <- list()\n",
    "for(celltype in c(\"K562\")){\n",
    "  message(celltype)\n",
    "  # inter_stats[[celltype]] <- list()\n",
    "  for(enh in c(\"En19\")){\n",
    "    message(enh)\n",
    "    # inter_stats[[celltype]][[enh]] <- list()\n",
    "    plot_temp <- fig3[which(fig3$enhancer==enh & fig3$celltype==celltype),]\n",
    "    if(nrow(plot_temp[which(plot_temp$project==\"motif\"),])==0) next\n",
    "    \n",
    "    ## Cycle through the sets that I need to plot along the x-axis and subset each plot_temp and skew_temp to include only that\n",
    "    ## Rename the columns to have the relevant column include only the immediately relevant columns: celltype, enhancer, silencer, ID, log2FoldChange, LogSkew and relevant binding\n",
    "    ## re-name columns to normalize the name of the binding column across all different sets.\n",
    "    ## Make sure to get the names to match up properly for xlab and legend title\n",
    "    \n",
    "    # for(inter in c(\"REST_RCOR_HDAC\", \"REST_MIER_EHMT2\")){\n",
    "    # message(inter)\n",
    "    # skew_levels <- c(\"-/-/-\",\"+/-/-\",\"+/+/-\",\"+/-/+\",\"+/+/+\",\"-/+/-\",\"-/-/+\",\"-/+/+\")\n",
    "    # l2fc_levels <- c(\"NegCtrl\",\"+/-/-\",\"+/+/-\",\"+/-/+\",\"+/+/+\",\"-/-/-\",\"-/+/-\",\"-/-/+\",\"-/+/+\")\n",
    "    \n",
    "    # if(inter==\"REST_RCOR_HDAC\"){\n",
    "    #   lab_name <- \"REST/RCOR/HDAC binding\"\n",
    "    # }\n",
    "    # if(inter==\"REST_MIER_EHMT2\"){\n",
    "    #   lab_name <- \"REST/MIER1/EHMT2\"\n",
    "    # }\n",
    "    \n",
    "    \n",
    "    l2fc_levels <- c(\"NegCtrl\",\"+/-/-/-/-\",\"+/+/-/-/-\",\"+/-/+/-/-\",\"+/-/-/+/-\",\"+/-/-/-/+\",\"+/+/+/-/-\",\"+/+/-/+/-\",\"+/+/-/-/+\",\"+/-/+/+/-\",\"+/-/+/-/+\",\"+/-/-/+/+\",\"+/-/+/+/+\",\"+/+/-/+/+\",\"+/+/+/-/+\",\"+/+/+/+/-\",\"+/+/+/+/+\",\n",
    "                     \"-/-/-/-/-\",\"-/+/-/-/-\",\"-/-/+/-/-\",\"-/-/-/+/-\",\"-/-/-/-/+\",\"-/+/+/-/-\",\"-/+/-/+/-\",\"-/+/-/-/+\",\"-/-/+/+/-\",\"-/-/+/-/+\",\"-/-/-/+/+\",\"-/-/+/+/+\",\"-/+/+/-/+\",\"-/+/-/+/+\",\"-/+/+/+/-\",\"-/+/+/+/+\")\n",
    "    lab_name <- \"REST/MIER/EHMT2/RCOR/HDAC2\"\n",
    "    \n",
    "    red_plot <- plot_temp[,c(\"ID\", \"enhancer\",\"silencer\", \"celltype\", \"log2FoldChange\", \"scr_l2fc\", \"LogSkew\", \"project\",\"all\")] #inter instead of all if redoing things\n",
    "    colnames(red_plot) <- c(colnames(red_plot)[1:8],\"binding\")\n",
    "    # skew_temp <- red_plot[complete.cases(red_plot$LogSkew),]\n",
    "    \n",
    "    red_plot$binding <- factor(red_plot$binding, levels = l2fc_levels)\n",
    "    \n",
    "    # skew_temp$binding <- factor(skew_temp$binding, levels = skew_levels)\n",
    "    \n",
    "    l2fc_min <- -2.5\n",
    "    l2fc_max <- 10.5\n",
    "    # skew_min <- -3\n",
    "    # skew_max <- 5.5\n",
    "    \n",
    "    neg_ctrl_med <- median(red_plot$log2FoldChange[which(red_plot$project==\"NegCtrl\")], na.rm = T)\n",
    "    \n",
    "    pdf(paste0(\"plots/tf_analysis/box_plots/neg_ctrl_motif_comp/\",enh,\"_\",celltype,\"_all_yrange.pdf\"), width=10, height = 5) #width=40/25.4, height=45/25.4)\n",
    "    print(ggplot(red_plot, aes(x=binding,y=log2FoldChange, fill=binding)) + geom_hline(yintercept = neg_ctrl_med,lwd=0.2) + geom_boxplot(outlier.shape = NA,lwd=0.2) + theme_light(base_size = 6) +\n",
    "            coord_cartesian(ylim = c(l2fc_min,l2fc_max)) + xlab(lab_name) + # scale_fill_manual(values=c(\"grey\",brewer_paired[c(1,3,5,7,2,4,6,8)])) + xlab(lab_name) +\n",
    "            theme(legend.position = \"none\",axis.text.x = element_text(angle = 45, hjust = 1)))\n",
    "    dev.off()\n",
    "    # pdf(paste0(\"plots/tf_analysis/box_plots/skew/\",enh,\"_\",celltype,\"_\",inter,\"_yrange.pdf\"), width=40/25.4, height=45/25.4)\n",
    "    # print(ggplot(skew_temp, aes(x=binding,y=LogSkew, fill=binding)) + geom_hline(yintercept = 0,lwd=0.2) + geom_boxplot(outlier.shape = NA, lwd=0.2) + theme_light(base_size = 6) +\n",
    "    #         coord_cartesian(ylim = c(skew_min,skew_max)) + scale_fill_manual(values=c(brewer_paired[c(2,1,3,5,7,4,6,8)])) + xlab(lab_name) +\n",
    "    #         theme(legend.position = \"none\",axis.text.x = element_text(angle = 45, hjust = 1)))\n",
    "    # dev.off()\n",
    "    \n",
    "    # }\n",
    "    \n",
    "    # message(\"l2fc sig\")\n",
    "    # inter_stats[[celltype]][[enh]][[inter]]$l2fc <- red_plot %>%\n",
    "    #   wilcox_test(formula = log2FoldChange ~ binding, ref.group = \"NegCtrl\") %>%\n",
    "    #   adjust_pvalue(method = \"BH\") %>%\n",
    "    #   add_significance()\n",
    "    # \n",
    "    # message(\"skew sig\")\n",
    "    # if(inter==\"CoREST\"){\n",
    "    #   inter_stats[[celltype]][[enh]][[inter]]$skew <- skew_temp %>%\n",
    "    #     wilcox_test(formula = LogSkew ~ binding, ref.group = \"-/-\") %>%\n",
    "    #     adjust_pvalue(method = \"BH\") %>%\n",
    "    #     add_significance()\n",
    "    # }\n",
    "    # if(inter!=\"CoREST\"){\n",
    "    #   inter_stats[[celltype]][[enh]][[inter]]$skew <- skew_temp %>%\n",
    "    #     wilcox_test(formula = LogSkew ~ binding, ref.group = \"-/-/-\") %>%\n",
    "    #     adjust_pvalue(method = \"BH\") %>%\n",
    "    #     add_significance()\n",
    "    # }\n",
    "  }\n",
    "}\n",
    "\n",
    "fig3$TRIP13 <- \"-\"\n",
    "fig3$PTTG1 <- \"-\"\n",
    "for(celltype in unique(fig3$celltype)){\n",
    "  message(celltype)\n",
    "  fig3$TRIP13[which(fig3$celltype==celltype & fig3$silencer %in% tf_sil_list[[celltype]]$TRIP13)] <- \"+\"\n",
    "  fig3$PTTG1[which(fig3$celltype==celltype & fig3$silencer %in% tf_sil_list[[celltype]]$PTTG1)] <- \"+\"\n",
    "  \n",
    "}\n",
    "\n",
    "fig3$score <- \"NoScore\"\n",
    "for(celltype in unique(fig3$celltype)){\n",
    "  message(celltype)\n",
    "  for(enh in enh_all){\n",
    "    message(enh)\n",
    "    for(relat in names(oligo_split_list[[celltype]][[enh]])){\n",
    "      fig3$score[which(fig3$silencer %in% oligo_split_list[[celltype]][[enh]][[relat]] & \n",
    "                         fig3$celltype==celltype & fig3$enhancer==enh)] <- relat\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "fig3$score[which(fig3$score==\"NoScore\" & fig3$project==\"motif\")] <- \"below\"\n",
    "\n",
    "\n",
    "\n",
    "for(celltype in unique(fig3$celltype)){\n",
    "  message(celltype)\n",
    "  for(enh in enh_all){\n",
    "    message(enh)\n",
    "    for(rel in c(\"above\",\"below\")){\n",
    "      message(rel)\n",
    "      plot_temp <- fig3[which(fig3$celltype==celltype & fig3$enhancer==enh & fig3$score %in% c(\"NoScore\",rel)),]\n",
    "      message(nrow(plot_temp))\n",
    "      for(inter in c(\"REST_RCOR_HDAC\", \"REST_MIER_EHMT2\")){\n",
    "        message(inter)\n",
    "        skew_levels <- c(\"-/-/-\",\"+/-/-\",\"+/+/-\",\"+/-/+\",\"+/+/+\",\"-/+/-\",\"-/-/+\",\"-/+/+\")\n",
    "        l2fc_levels <- c(\"NegCtrl\",\"+/-/-\",\"+/+/-\",\"+/-/+\",\"+/+/+\",\"-/-/-\",\"-/+/-\",\"-/-/+\",\"-/+/+\")\n",
    "        \n",
    "        if(inter==\"REST_RCOR_HDAC\"){\n",
    "          lab_name <- \"REST/RCOR/HDAC binding\"\n",
    "        }\n",
    "        if(inter==\"REST_MIER_EHMT2\"){\n",
    "          lab_name <- \"REST/MIER1/EHMT2\"\n",
    "        }\n",
    "        \n",
    "        red_plot <- plot_temp[,c(\"ID\", \"enhancer\",\"silencer\", \"celltype\", \"log2FoldChange\", \"scr_l2fc\", \"LogSkew\", \"project\",inter)]\n",
    "        message(nrow(red_plot))\n",
    "        colnames(red_plot) <- c(\"ID\", \"enhancer\",\"silencer\", \"celltype\", \"log2FoldChange\", \"scr_l2fc\", \"LogSkew\", \"project\",\"binding\")\n",
    "        skew_temp <- red_plot[complete.cases(red_plot$LogSkew),]\n",
    "        \n",
    "        red_plot$binding <- factor(red_plot$binding, levels = l2fc_levels)\n",
    "        \n",
    "        skew_temp$binding <- factor(skew_temp$binding, levels = skew_levels)\n",
    "        \n",
    "        # l2fc_min <- -2.5\n",
    "        # l2fc_max <- 10.5\n",
    "        # skew_min <- -3\n",
    "        # skew_max <- 5.5\n",
    "        \n",
    "        neg_ctrl_med <- median(red_plot$log2FoldChange[which(red_plot$project==\"NegCtrl\")], na.rm = T)\n",
    "        \n",
    "        pdf(paste0(\"plots/tf_analysis/box_plots/neg_ctrl_motif_comp/fig_S8_above_below/\",enh,\"_\",celltype,\"_\",inter,\"_\",rel,\".pdf\"), width=40/25.4, height=45/25.4)\n",
    "        print(ggplot(red_plot, aes(x=binding,y=log2FoldChange, fill=binding)) + geom_hline(yintercept = neg_ctrl_med,lwd=0.2) + geom_boxplot(outlier.shape = NA,lwd=0.2) + theme_light(base_size = 6) +\n",
    "                # coord_cartesian(ylim = c(l2fc_min,l2fc_max)) + \n",
    "                scale_fill_manual(values=c(\"grey\",brewer_paired[c(1,3,5,7,2,4,6,8)])) + xlab(lab_name) +\n",
    "                theme(legend.position = \"none\",axis.text.x = element_text(angle = 45, hjust = 1)))\n",
    "        dev.off()\n",
    "        pdf(paste0(\"plots/tf_analysis/box_plots/skew/fig_s8_above_below/\",enh,\"_\",celltype,\"_\",inter,\"_\",rel,\".pdf\"), width=40/25.4, height=45/25.4)\n",
    "        print(ggplot(skew_temp, aes(x=binding,y=LogSkew, fill=binding)) + geom_hline(yintercept = 0,lwd=0.2) + geom_boxplot(outlier.shape = NA, lwd=0.2) + theme_light(base_size = 6) +\n",
    "                # coord_cartesian(ylim = c(skew_min,skew_max)) + \n",
    "                scale_fill_manual(values=c(brewer_paired[c(2,1,3,5,7,4,6,8)])) + xlab(lab_name) +\n",
    "                theme(legend.position = \"none\",axis.text.x = element_text(angle = 45, hjust = 1)))\n",
    "        dev.off()\n",
    "        # \n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "fig3$ZNF644 <- \"-\"\n",
    "\n",
    "for(celltype in unique(fig3$celltype)){\n",
    "  message(celltype)\n",
    "  fig3$ZNF644[which(fig3$celltype==celltype & fig3$silencer %in% tf_sil_list[[celltype]]$ZNF644)] <- \"+\"\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa520d8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Figure 5G\n",
    "## Make density plots comparing the minor allele frequency inside and outside strong motifs\n",
    "\n",
    "emvar_comp <- data.frame()\n",
    "for(celltype in names(emVAR_all)){\n",
    "  keep <- emVAR_all[[celltype]][which(emVAR_all[[celltype]]$SNP %in% SNP_incl & !is.na(emVAR_all[[celltype]]$ref_allele)), c(\"ID\",\"SNP\",\"LogSkew\",\"Skew_logFDR\")]\n",
    "  keep$celltype <- celltype\n",
    "  emvar_comp <- rbind(emvar_comp, keep)\n",
    "}\n",
    "\n",
    "## Compile scores for ref and alt alleles into one data frame identifying which have the SNP inside and outside the center window\n",
    "maf_rare_ref <- read.delim(\"meme-suite/maf_rare_ref_fimo.txt\", stringsAsFactors = F)\n",
    "maf_rare_alt <- read.delim(\"meme-suite/maf_rare_alt_fimo.txt\", stringsAsFactors = F)\n",
    "\n",
    "head(maf_rare_alt)\n",
    "head(maf_rare_ref)\n",
    "\n",
    "maf_rare_alt <- maf_rare_alt[order(maf_rare_alt$sequence.name),]\n",
    "maf_rare_ref <- maf_rare_ref[order(maf_rare_ref$sequence.name),]\n",
    "\n",
    "maf_rare_alt <- maf_rare_alt[maf_rare_alt$sequence.name!=\"\",]\n",
    "maf_rare_ref <- maf_rare_ref[maf_rare_ref$sequence.name!=\"\",]\n",
    "\n",
    "maf_rare_alt <- maf_rare_alt[which(maf_rare_alt$X.pattern.name==\"MA0138.2\"),]\n",
    "maf_rare_ref <- maf_rare_ref[which(maf_rare_ref$X.pattern.name==\"MA0138.2\"),]\n",
    "\n",
    "maf_rare_alt$loc <- \"outside\"\n",
    "maf_rare_ref$loc <- \"outside\"\n",
    "\n",
    "maf_rare_alt$loc[which(maf_rare_alt$start>90 & maf_rare_alt$stop < 112)] <- \"inside\"\n",
    "maf_rare_ref$loc[which(maf_rare_ref$start>90 & maf_rare_ref$stop < 112)] <- \"inside\"\n",
    "\n",
    "mra_split <- colsplit(maf_rare_alt$sequence.name,\":\",c(\"chr\",\"pos\",\"ref\",\"alt\",\"allele\",\"window\"))\n",
    "mrr_split <- colsplit(maf_rare_ref$sequence.name,\":\",c(\"chr\",\"pos\",\"ref\",\"alt\",\"allele\",\"window\"))\n",
    "\n",
    "mra_window <- colsplit(mra_split$window, \"P\", c(\"w\",\"num\"))\n",
    "mrr_window <- colsplit(mrr_split$window, \"P\", c(\"w\",\"num\"))\n",
    "\n",
    "maf_rare_alt$SNP <- paste(mra_split$chr,mra_split$pos,mra_split$ref,mra_split$alt, sep=\":\")\n",
    "maf_rare_ref$SNP <- paste(mrr_split$chr,mrr_split$pos,mrr_split$ref,mrr_split$alt, sep=\":\")\n",
    "\n",
    "maf_rare_alt$num <- mra_window$num\n",
    "maf_rare_ref$num <- mrr_window$num\n",
    "\n",
    "maf_rare_scores <- data.frame()\n",
    "maf_rare_snps <- c(unique(maf_rare_alt$SNP), unique(maf_rare_ref$SNP))\n",
    "for(snp in unique(maf_rare_snps)){\n",
    "  message(snp)\n",
    "  alt_temp <- maf_rare_alt[which(maf_rare_alt$SNP==snp),]\n",
    "  ref_temp <- maf_rare_ref[which(maf_rare_ref$SNP==snp),]\n",
    "  \n",
    "  # if(nrow(alt_temp==0)){\n",
    "  #   score_a <- 0\n",
    "  #   loc_a <- NA\n",
    "  # }\n",
    "  # if(nrow(ref_temp==0)){\n",
    "  #   message(\"no ref\")\n",
    "  #   score_r <- 0\n",
    "  #   loc_r <- NA\n",
    "  # }\n",
    "  score_a <- 0\n",
    "  loc_a <- NA\n",
    "  score_r <- 0\n",
    "  loc_r <- NA\n",
    "  \n",
    "  if(nrow(alt_temp)==1){\n",
    "    score_a <- alt_temp$score\n",
    "    loc_a <- alt_temp$loc\n",
    "  }\n",
    "  if(nrow(ref_temp)==1){\n",
    "    score_r <- ref_temp$score\n",
    "    loc_r <- ref_temp$loc\n",
    "  }\n",
    "  \n",
    "  if(nrow(alt_temp) > 1){\n",
    "    message(\"multiple alt\")\n",
    "    if(\"inside\" %in% alt_temp$loc){\n",
    "      alt_temp <- alt_temp[which(alt_temp$loc==\"inside\"),]\n",
    "      if(nrow(alt_temp)>1){\n",
    "        pos_diff <- abs(100 - unique(alt_temp$num))\n",
    "        alt_temp <- alt_temp[which(pos_diff==min(pos_diff)),]\n",
    "      }\n",
    "      score_a <- alt_temp$score\n",
    "      loc_a <- alt_temp$loc\n",
    "    }\n",
    "    else{\n",
    "      # message(\"mult outside\")\n",
    "      loc_a <- \"outside\"\n",
    "      score_a <- mean(alt_temp$score, na.rm=T)\n",
    "    }\n",
    "  }\n",
    "  if(nrow(ref_temp) > 1){\n",
    "    # message(\"multiple ref\")\n",
    "    if(\"inside\" %in% ref_temp$loc){\n",
    "      ref_temp <- ref_temp[which(ref_temp$loc==\"inside\"),]\n",
    "      if(nrow(ref_temp)>1){\n",
    "        pos_diff <- abs(100 - unique(ref_temp$num))\n",
    "        ref_temp <- ref_temp[which(pos_diff==min(pos_diff)),]\n",
    "      }\n",
    "      score_r <- ref_temp$score\n",
    "      loc_r <- ref_temp$loc\n",
    "    }\n",
    "    else{\n",
    "      # message(\"mult outside\")\n",
    "      loc_r <- \"outside\"\n",
    "      score_r <- mean(ref_temp$score, na.rm=T)\n",
    "    }\n",
    "  }\n",
    "\n",
    "  a_tmp <- data.frame(snp, score_a, loc_a, score_r, loc_r)\n",
    "  maf_rare_scores <- rbind(maf_rare_scores, a_tmp)\n",
    "}\n",
    "\n",
    "emvar_comp <- merge(emvar_comp[,1:10], maf_rare_scores, by.x=\"SNP\", by.y=\"snp\", all.x=T)\n",
    "\n",
    "emvar_comp$DBS <- emvar_comp$score_a - emvar_comp$score_r\n",
    "\n",
    "## Incorporate the population minor allele frequency (MAF) for each oligo and incorporate into the df\n",
    "\n",
    "maf1_pop_af <- read.delim(\"fasta/maf1_AF_1KGp3_vcf.txt\", header = F, sep = \"\\t\", stringsAsFactors = F)\n",
    "rare_pop_af <- read.delim(\"fasta/rare_AF_1KGp3_vcf.txt\", header = F, sep = \"\\t\", stringsAsFactors = F)\n",
    "\n",
    "dim(maf1_pop_af)\n",
    "\n",
    "colnames(maf1_pop_af) <- c(\"chr\", \"pos\",\"rsid\", \"ref\",\"alt\",\"qual\",\"filter\",\"info\")\n",
    "colnames(rare_pop_af) <- c(\"chr\", \"pos\",\"rsid\", \"ref\",\"alt\",\"qual\",\"filter\",\"info\")\n",
    "\n",
    "maf1_info_split <- colsplit(maf1_pop_af$info, \"\\\\;\", names = c(\"AC\", \"AF\",\"AN\",\"NS\",\"DP\",\"EAS\",\"AMR\",\"AFR\",\"EUR\",\"SAS\",\"AA\",\"VT\"))\n",
    "rare_info_split <- colsplit(rare_pop_af$info, \"\\\\;\", names = c(\"AC\", \"AF\",\"AN\",\"NS\",\"DP\",\"EAS\",\"AMR\",\"AFR\",\"EUR\",\"SAS\",\"AA\",\"VT\"))\n",
    "head(maf1_info_split)\n",
    "\n",
    "maf1_EAS <- colsplit(maf1_info_split$EAS, \"=\", names=c(\"name\",\"EAS_AF\"))\n",
    "maf1_EUR <- colsplit(maf1_info_split$EUR, \"=\", names=c(\"name\",\"EUR_AF\"))\n",
    "maf1_AFR <- colsplit(maf1_info_split$AFR, \"=\", names=c(\"name\",\"AFR_AF\"))\n",
    "\n",
    "rare_EAS <- colsplit(rare_info_split$EAS, \"=\", names=c(\"name\",\"EAS_AF\"))\n",
    "rare_EUR <- colsplit(rare_info_split$EUR, \"=\", names=c(\"name\",\"EUR_AF\"))\n",
    "rare_AFR <- colsplit(rare_info_split$AFR, \"=\", names=c(\"name\",\"AFR_AF\"))\n",
    "\n",
    "maf1_pop_af <- cbind(maf1_pop_af[,1:7], maf1_EAS$EAS_AF, maf1_EUR$EUR_AF, maf1_AFR$AFR_AF)\n",
    "rare_pop_af <- cbind(rare_pop_af[,1:7], rare_EAS$EAS_AF, rare_EUR$EUR_AF, rare_AFR$AFR_AF)\n",
    "\n",
    "colnames(maf1_pop_af) <- c(\"chr\", \"pos\",\"rsid\", \"ref\",\"alt\",\"qual\",\"filter\",\"EAS_AF\", \"EUR_AF\",\"AFR_AF\")\n",
    "colnames(rare_pop_af) <- c(\"chr\", \"pos\",\"rsid\", \"ref\",\"alt\",\"qual\",\"filter\",\"EAS_AF\", \"EUR_AF\", \"AFR_AF\")\n",
    "\n",
    "\n",
    "maf1_pop_af$SNP <- paste(maf1_pop_af$chr,maf1_pop_af$pos,maf1_pop_af$ref,maf1_pop_af$alt,sep = \":\")\n",
    "rare_pop_af$SNP <- paste(rare_pop_af$chr,rare_pop_af$pos,rare_pop_af$ref,rare_pop_af$alt,sep = \":\")\n",
    "\n",
    "install.packages(\"taRifx\")\n",
    "library(taRifx)\n",
    "maf1_pop_af <- remove.factors(maf1_pop_af)\n",
    "maf1_pop_af$EAS_AF <- as.numeric(maf1_pop_af$EAS_AF)\n",
    "maf1_pop_af$max_AF <- pmax(as.numeric(maf1_pop_af$EAS_AF),as.numeric(maf1_pop_af$EUR_AF), as.numeric(maf1_pop_af$AFR_AF), na.rm = T)\n",
    "summary(maf1_pop_af)\n",
    "\n",
    "rare_pop_af <- remove.factors(rare_pop_af)\n",
    "rare_pop_af$EAS_AF <- as.numeric(rare_pop_af$EAS_AF)\n",
    "rare_pop_af$EUR_AF <- as.numeric(rare_pop_af$EUR_AF)\n",
    "rare_pop_af$AFR_AF <- as.numeric(rare_pop_af$AFR_AF)\n",
    "rare_pop_af$max_AF <- pmax(as.numeric(rare_pop_af$EAS_AF), as.numeric(rare_pop_af$EUR_AF), as.numeric(rare_pop_af$AFR_AF), na.rm = T)\n",
    "summary(rare_pop_af)\n",
    "\n",
    "rare_pop_af_og[rare_AFR$AFR_AF==2504,]\n",
    "\n",
    "maf1_SNP <- RESTscreen_attributes$SNP[RESTscreen_attributes$project==\"maf1\"]\n",
    "rare_SNP <- RESTscreen_attributes$SNP[RESTscreen_attributes$project==\"rare\"]\n",
    "\n",
    "SNP_incl <- c(maf1_SNP,rare_SNP)\n",
    "pop_AF <- rbind(rare_pop_af, maf1_pop_af)\n",
    "pop_AF <- pop_AF[which(pop_AF$SNP %in% SNP_incl),]\n",
    "\n",
    "emvar_comp$maf <- \">1\"\n",
    "emvar_comp$maf[which(emvar_comp$SNP %in% pop_AF$SNP[which(pop_AF$max_AF < 0.01)])] <- \"<1\"\n",
    "emvar_comp$maf[which(emvar_comp$SNP %in% pop_AF$SNP[which(pop_AF$max_AF > 0.05)])] <- \">5\"\n",
    "emvar_comp$maf[which(emvar_comp$SNP %in% pop_AF$SNP[which(pop_AF$max_AF > 0.1)])] <- \">10\"\n",
    "\n",
    "emvar_comp$maf <- factor(emvar_comp$maf, levels = c(\"<1\",\">1\",\">5\"))\n",
    "\n",
    "emvar_comp$window <- \"outside\"\n",
    "emvar_comp$window[which(emvar_comp$loc_a==\"inside\" & emvar_comp$loc_r==\"inside\")] <- \"inside\"\n",
    "\n",
    "comp_sil_split <- colsplit(emvar_comp$silencer,\":wP\",c(\"sil_allele\",\"wP\"))\n",
    "# emvar_comp$window[which(emvar_comp$loc_a!=emvar_comp$loc_r)] <- ifelse(comp_sil_split$wP[which(emvar_comp$loc_a!=emvar_comp$loc_r)] < 112 & comp_sil_split$wP[which(emvar_comp$loc_a!=emvar_comp$loc_r)] > 90, \"inside\",\"outside\")\n",
    "emvar_comp$window <- ifelse(comp_sil_split$wP < 112 & comp_sil_split$wP > 90, \"inside\",\"outside\")\n",
    "\n",
    "emvar_comp$rel <- \"below\"\n",
    "emvar_comp$rel[which(emvar_comp$score_a >= 20.86 | emvar_comp$score_r >= 20.86)] <- \"above\"\n",
    "\n",
    "## Make the plots and calculate statistics by wilcox test comparing window position and relative score\n",
    "\n",
    "emvar_comp_stat <- list()\n",
    "for(celltype in unique(emvar_comp$celltype)){\n",
    "  emvar_comp_stat[[celltype]] <- list()\n",
    "  message(celltype)\n",
    "  for(enh in enh_all){\n",
    "    emvar_comp_stat[[celltype]][[enh]] <- list()\n",
    "    message(enh)\n",
    "    for(rel_pos in c(\"above\",\"below\")){\n",
    "      emvar_comp_stat[[celltype]][[enh]][[rel_pos]] <- list()\n",
    "      message(rel_pos)\n",
    "      for(win in c(\"inside\",\"outside\")){\n",
    "        message(win)\n",
    "        to_plot <- emvar_comp[which(emvar_comp$celltype==celltype & emvar_comp$enhancer==enh & emvar_comp$maf!=\">1\" & emvar_comp$window==win & emvar_comp$rel==rel_pos),]\n",
    "        \n",
    "        pdf(paste0(\"plots/minor_allele_frequency_analysis/density/\",celltype,\"_\",enh,\"_maf_density_\",rel_pos,\"_\",win,\".pdf\"), width=40/25.4, height=40/25.4)\n",
    "        print(ggplot(to_plot, aes(x=LogSkew, group=maf, fill=maf)) + geom_density(adjust=1.5, alpha=0.2, lwd=0.1) + xlim(-5,5) +\n",
    "                theme_minimal(base_size = 6) + theme(legend.position = \"none\"))\n",
    "        dev.off()\n",
    "        \n",
    "        to_stat <- remove.factors(to_plot)\n",
    "        stat_temp <- wilcox_test(to_stat, LogSkew~maf, ref.group = \"<1\", p.adjust.method = \"BH\")\n",
    "        emvar_comp_stat[[celltype]][[enh]][[rel_pos]][[win]] <- stat_temp\n",
    "        \n",
    "        # pdf(paste0(\"plots/minor_allele_frequency_analysis/density/\",celltype,\"_\",enh,\"_maf_density_\",rel_pos,\"_\",win,\".pdf\"), width=40/25.4, height=40/25.4)\n",
    "        # print(ggplot(to_plot, aes(x=LogSkew, group=maf3b, fill=maf3b)) + geom_density(adjust=1.5, alpha=0.2, lwd=0.1) +\n",
    "        #         theme_minimal(base_size = 6) + theme(legend.position = \"none\"))\n",
    "        # dev.off()\n",
    "        \n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "emvar_comp_stat_df <- data.frame()\n",
    "for(celltype in names(emvar_comp_stat)){\n",
    "  message(celltype)\n",
    "  for(enhancer in names(emvar_comp_stat[[celltype]])){\n",
    "    message(enhancer)\n",
    "    for(relativePosition in names(emvar_comp_stat[[celltype]][[enhancer]])){\n",
    "      message(relativePosition)\n",
    "      for(window in names(emvar_comp_stat[[celltype]][[enhancer]][[relativePosition]])){\n",
    "        message(window)\n",
    "        stat_temp <- data.frame(emvar_comp_stat[[celltype]][[enhancer]][[relativePosition]][[window]])\n",
    "        stat_temp$celltype <- celltype\n",
    "        stat_temp$enhancer <- enhancer\n",
    "        stat_temp$relativePosition <- relativePosition\n",
    "        stat_temp$window <- window\n",
    "        emvar_comp_stat_df <- rbind(emvar_comp_stat_df, stat_temp)\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "emvar_comp_stat_df$ks_p <- NA\n",
    "for(celltype in unique(emvar_comp_stat_df$celltype)){\n",
    "  message(celltype)\n",
    "  for(enh in unique(emvar_comp_stat_df$enhancer)){\n",
    "    message(enh)\n",
    "    for(rel in unique(emvar_comp_stat_df$relativePosition)){\n",
    "      message(rel)\n",
    "      for(win in unique(emvar_comp_stat_df$window)){\n",
    "        message(win)\n",
    "        g1 <- emvar_comp$LogSkew[which(emvar_comp$celltype==celltype & emvar_comp$enhancer==enh & emvar_comp$rel==rel & emvar_comp$window==win & emvar_comp$maf3b==\"<1\")]\n",
    "        g2 <- emvar_comp$LogSkew[which(emvar_comp$celltype==celltype & emvar_comp$enhancer==enh & emvar_comp$rel==rel & emvar_comp$window==win & emvar_comp$maf3b==\">5\")]\n",
    "        \n",
    "        emvar_comp_stat_df$ks_p[which(emvar_comp_stat_df$celltype==celltype & emvar_comp_stat_df$enhancer==enh & emvar_comp_stat_df$relativePosition==rel & emvar_comp_stat_df$window==win)] <- ks.test(g1,g2)$p.value\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
